{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "edf75ab0-5449-4cd9-81e0-a80793fbbd1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "import numpy as np\n",
    "from sgRNAalign_util import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb668a83-0a8c-4e64-9fe1-3a62f88c31e8",
   "metadata": {},
   "source": [
    "### Prepare sgRNA alignment file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8f34e764-ed85-4df6-9d79-5fc6d8b46fdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading from 'genome/GWCD4CRISPRi_targeting_sgRNA_list.csv'...\n",
      "Successfully converted 25512 records.\n",
      "FASTA file saved as 'genome/GWCD4CRISPRi_targeting_sgRNA_list.fa'\n"
     ]
    }
   ],
   "source": [
    "convert_csv_to_fasta('genome/GWCD4CRISPRi_targeting_sgRNA_list.csv', 'genome/GWCD4CRISPRi_targeting_sgRNA_list.fa')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "675f3490-47d4-4794-ab6c-076d956f7e20",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "bowtie2 -x /Users/rzhu/Gladstone Dropbox/Ronghui Zhu/GRNPerturbSeq/2_files/hg38_genome/hg38 \\\n",
    "        -f genome/GWCD4CRISPRi_targeting_sgRNA_list.fa \\\n",
    "        -S results/GWCD4CRISPRi_targeting_sgRNA_alignment.sam \\\n",
    "        --end-to-end \\\n",
    "        -D 20 \\\n",
    "        -R 3 \\\n",
    "        -L 5 \\\n",
    "        -N 0 \\\n",
    "        -k 100 \\\n",
    "        -i S,1,0.50 \\\n",
    "        --score-min \"C,0,0\" \\\n",
    "        --no-1mm-upfront \\\n",
    "        --no-unal \\\n",
    "        --un results/unaligned_targeting_sgrnas.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea05322d-9781-4ab9-b100-3e00775a3919",
   "metadata": {},
   "source": [
    "25512 reads; of these:\n",
    "  25512 (100.00%) were unpaired; of these:\n",
    "    7 (0.03%) aligned 0 times\n",
    "    23993 (94.05%) aligned exactly 1 time\n",
    "    1512 (5.93%) aligned >1 times\n",
    "99.97% overall alignment rate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6b6fa5f-5469-4305-a421-a388d518ab2c",
   "metadata": {},
   "source": [
    "### Extract alignment info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a570970e-928c-4e14-ace8-89e1c603e705",
   "metadata": {},
   "outputs": [],
   "source": [
    "alignment_df = sam_to_dataframe('results/GWCD4CRISPRi_targeting_sgRNA_alignment.sam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7f09e14b-bcbe-4fe8-a821-4109312a4377",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get uniquely aligned sgRNAs\n",
    "sgrna_counts = alignment_df['sgRNA'].value_counts()\n",
    "unique_sgrna_ids = sgrna_counts[sgrna_counts == 1].index\n",
    "unique_sgrna_df = alignment_df[alignment_df['sgRNA'].isin(unique_sgrna_ids)].copy()\n",
    "unique_sgrna_df['note'] = 'unique alignment'\n",
    "unique_sgrna_df['PAM'] = True\n",
    "unique_sgrna_df['flag'] = False\n",
    "unique_sgrna_df = unique_sgrna_df.reset_index(drop=True)\n",
    "\n",
    "# Get sgRNAs with multi alignments\n",
    "sgrna_counts = alignment_df['sgRNA'].value_counts()\n",
    "multi_sgrna_ids = sgrna_counts[sgrna_counts > 1].index\n",
    "multi_sgrna_df = alignment_df[alignment_df['sgRNA'].isin(multi_sgrna_ids)].copy()\n",
    "multi_sgrna_df['note'] = 'multi alignment'\n",
    "multi_sgrna_df['PAM'] = True\n",
    "multi_sgrna_df = multi_sgrna_df.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "9f78e07a-7f66-4dcf-b68d-f3c486702db5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1512"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(multi_sgrna_df.sgRNA.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "fc720cff-71da-4037-a485-a925ff424efa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23993"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unique_sgrna_df.sgRNA.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48d685ba-94b6-4573-8034-9b5633a69f52",
   "metadata": {},
   "source": [
    "### First filter by whether there are PAM sequences after protospacer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b1880f53-1aad-411b-9420-e3ad34ec5c52",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyfaidx import Fasta\n",
    "\n",
    "# Load the reference genome\n",
    "try:\n",
    "    genome = Fasta('genome/hg38.fa')\n",
    "except FileNotFoundError:\n",
    "    print(\"Error: hg38.fa not found. Please download the reference genome.\")\n",
    "    exit()\n",
    "\n",
    "# Flag uniquely aligned sgRNA that has no PAM\n",
    "for i in range(len(unique_sgrna_df)):\n",
    "    # Define the genomic coordinates\n",
    "    chromosome = unique_sgrna_df.loc[i, 'chromosome']\n",
    "    pos = unique_sgrna_df.loc[i, 'pos']\n",
    "    if unique_sgrna_df.loc[i, 'strand']=='+':\n",
    "        start_position = pos+19\n",
    "        end_position = start_position+3\n",
    "        sequence = genome[chromosome][start_position:end_position]\n",
    "    else:\n",
    "        start_position = pos-3\n",
    "        end_position = pos\n",
    "        sequence = genome[chromosome][start_position:end_position]\n",
    "        sequence = sequence.reverse.complement\n",
    "    \n",
    "    if (sequence.seq[1:]!='GG') and (sequence.seq[1:]!='gg') and (sequence.seq[1:]!='Gg' and (sequence.seq[1:]!='gG')):\n",
    "        unique_sgrna_df.loc[i, 'PAM'] = False\n",
    "        unique_sgrna_df.loc[i, 'flag'] = True\n",
    "\n",
    "# For multi aligned sgRNA, only include those with PAM\n",
    "temp_df = multi_sgrna_df.copy()\n",
    "for i in range(len(multi_sgrna_df)):\n",
    "    chromosome = temp_df.loc[i, 'chromosome']\n",
    "    pos = temp_df.loc[i, 'pos']\n",
    "    if temp_df.loc[i, 'strand']=='+':\n",
    "        start_position = pos+19\n",
    "        end_position = start_position+3\n",
    "        sequence = genome[chromosome][start_position:end_position]\n",
    "    else:\n",
    "        start_position = pos-3\n",
    "        end_position = pos\n",
    "        sequence = genome[chromosome][start_position:end_position]\n",
    "        sequence = sequence.reverse.complement\n",
    "    \n",
    "    if (sequence.seq[1:]!='GG') and (sequence.seq[1:]!='gg') and (sequence.seq[1:]!='Gg' and (sequence.seq[1:]!='gG')):\n",
    "        temp_df.drop(i, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3cf45705-818b-49ae-aa37-1778a0e76505",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NBPF14-2'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# One of the multi alignment does not have PAM at all\n",
    "set(multi_sgrna_df.sgRNA.unique()) - set(temp_df.sgRNA.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1daf8faa-ab22-4bbf-8dbb-5ca82ba635a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1018"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get more uniquely aligned sgRNA\n",
    "sgrna_counts = temp_df['sgRNA'].value_counts()\n",
    "unique_sgrna_ids = sgrna_counts[sgrna_counts == 1].index\n",
    "unique_sgrna_df_from_multi_after_PAMfilter = temp_df[temp_df['sgRNA'].isin(unique_sgrna_ids)].copy()\n",
    "unique_sgrna_df_from_multi_after_PAMfilter['note'] = 'unique alignment'\n",
    "unique_sgrna_df_from_multi_after_PAMfilter['flag'] = False\n",
    "\n",
    "# Get multi-aligned sgRNA after PAM filtering\n",
    "sgrna_counts = temp_df['sgRNA'].value_counts()\n",
    "multi_sgrna_ids = sgrna_counts[sgrna_counts > 1].index\n",
    "multi_sgrna_df_after_PAMfilter = temp_df[temp_df['sgRNA'].isin(multi_sgrna_ids)].copy()\n",
    "multi_sgrna_df.loc[multi_sgrna_df.sgRNA=='NBPF14-2', 'PAM'] = False\n",
    "multi_sgrna_df_after_PAMfilter = pd.concat([multi_sgrna_df_after_PAMfilter, multi_sgrna_df[multi_sgrna_df.sgRNA=='NBPF14-2']])\n",
    "multi_sgrna_df_after_PAMfilter = multi_sgrna_df_after_PAMfilter.reset_index(drop=True)\n",
    "len(multi_sgrna_df_after_PAMfilter.sgRNA.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa9a6ca6-3454-4b6e-a2eb-16cbe56257eb",
   "metadata": {},
   "source": [
    "### Rest of multi alignments, filter by whether there are genes nearby (30kb) or within a gene body"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0537f57f-6261-4495-a0d6-f00c8f9c0c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_df_subset = pd.read_pickle('genome/genes_df_subset_for_sgRNA_annotation.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6ec7d092-2cf8-4992-964d-7d8a34b31134",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check nearby gene\n",
    "multi_sgrna_df_after_PAMfilter['gene_nearby'] = multi_sgrna_df_after_PAMfilter.apply(\n",
    "    is_gene_nearby, \n",
    "    axis=1, \n",
    "    args=(genes_df_subset,)\n",
    ")\n",
    "\n",
    "multi_sgrna_df_after_genePAMfilter = multi_sgrna_df_after_PAMfilter[multi_sgrna_df_after_PAMfilter.gene_nearby].copy()\n",
    "multi_sgrna_df_after_genePAMfilter = multi_sgrna_df_after_genePAMfilter.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2a909b29-1b08-44d4-8fbf-1f0e184492b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if there is any multi alignment sgRNA that have no nearby genes\n",
    "set(multi_sgrna_df_after_PAMfilter.sgRNA.unique()) - set(multi_sgrna_df_after_genePAMfilter.sgRNA.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ee9b0ca9-0acb-4b97-a5a7-480b13a6faa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get more uniquely aligned sgRNA after nearby gene filtering\n",
    "sgrna_counts = multi_sgrna_df_after_genePAMfilter['sgRNA'].value_counts()\n",
    "unique_sgrna_ids = sgrna_counts[sgrna_counts == 1].index\n",
    "unique_sgrna_df_from_multi_after_genePAMfilter = multi_sgrna_df_after_genePAMfilter[multi_sgrna_df_after_genePAMfilter['sgRNA'].isin(unique_sgrna_ids)].copy()\n",
    "unique_sgrna_df_from_multi_after_genePAMfilter['note'] = 'unique alignment'\n",
    "unique_sgrna_df_from_multi_after_genePAMfilter['flag'] = False\n",
    "\n",
    "# Get multi-aligned sgRNA after nearby gene filtering\n",
    "sgrna_counts = multi_sgrna_df_after_genePAMfilter['sgRNA'].value_counts()\n",
    "multi_sgrna_ids = sgrna_counts[sgrna_counts > 1].index\n",
    "multi_sgrna_df_filtered = multi_sgrna_df_after_genePAMfilter[multi_sgrna_df_after_genePAMfilter['sgRNA'].isin(multi_sgrna_ids)].copy()\n",
    "multi_sgrna_df_filtered = multi_sgrna_df_filtered.reset_index(drop=True)\n",
    "multi_sgrna_df_filtered['flag'] = 'True'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "58a1399b-13b8-419b-96ad-c43e10c8a9f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine all uniquely aligned sgRNA\n",
    "unique_sgrna_df_all = pd.concat([unique_sgrna_df, \n",
    "                                 unique_sgrna_df_from_multi_after_PAMfilter, \n",
    "                                 unique_sgrna_df_from_multi_after_genePAMfilter])\n",
    "unique_sgrna_df_all.drop(columns='gene_nearby', inplace=True)\n",
    "unique_sgrna_df_all = unique_sgrna_df_all.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "12cda4dc-f5b3-4399-8930-75f2b4293ee3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "238"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(multi_sgrna_df_filtered.sgRNA.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ce561006-1cfe-4151-b863-e04295c176ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25267"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unique_sgrna_df_all.sgRNA.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b9bd6fe-8360-48e6-a94b-7f699996fb32",
   "metadata": {},
   "source": [
    "### Annotate uniquely aligned sgRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4594220b-158d-46cc-99cb-1f2643818a95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add sgRNA information\n",
    "sgRNA = pd.read_pickle('genome/GWCD4CRISPRi_sgRNA_list_updated.pkl')\n",
    "df1 = unique_sgrna_df_all[['sgRNA', 'chromosome', 'pos', 'seq', 'strand', 'PAM', 'note', 'flag']].copy()\n",
    "df1.rename(columns={'seq': 'seq_last19bp'}, inplace=True)\n",
    "df2 = sgRNA[['name', 'sgRNA', 'target_gene_name_from_sgRNA', 'gene_id', 'corrected_target_gene_name']].copy()\n",
    "df2.rename(columns={\n",
    "    'sgRNA': 'seq',\n",
    "    'gene_id': 'designed_target_gene_id',\n",
    "    'corrected_target_gene_name': 'designed_target_gene_name'\n",
    "}, inplace=True)\n",
    "merged_df = pd.merge(\n",
    "    df1,\n",
    "    df2,\n",
    "    left_on='sgRNA',\n",
    "    right_on='name',\n",
    "    how='left'  # 'left' merge keeps all records from the first DataFrame\n",
    ")\n",
    "merged_df.drop(columns='name', inplace=True)\n",
    "unique_sgrna_df_all = merged_df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f6df1b7-62f4-4007-b1f1-e1bf2e9c3428",
   "metadata": {},
   "source": [
    "#### Check whether near target genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1be5f5e1-2d4c-4898-b2dd-5c9258029c5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check whether target gene is within certain distance (2kb)\n",
    "unique_sgrna_df_all['chromosome_norm'] = unique_sgrna_df_all['chromosome'].str.replace('chr', '')\n",
    "genes_df_subset['chromosome_norm'] = genes_df_subset['chromosome'].str.replace('chr', '')\n",
    "merged_df = pd.merge(\n",
    "    unique_sgrna_df_all,\n",
    "    genes_df_subset[['gene_id', 'gene_name', 'chromosome_norm', 'tss', 'cds', 'gene_start', 'gene_end']],\n",
    "    left_on=['designed_target_gene_id', 'chromosome_norm'],\n",
    "    right_on=['gene_id', 'chromosome_norm'],\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "primary_target_cols = ['target_gene_id', 'target_gene_name', 'distance_to_closest_target_tss']\n",
    "merged_df[primary_target_cols] = merged_df.apply(find_closest_target_info, axis=1)\n",
    "\n",
    "match_condition = merged_df['designed_target_gene_id']==merged_df['target_gene_id']\n",
    "merged_df.loc[match_condition, 'note'] = 'unique alignment, aligned to target gene'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dec80f17-5af2-4a75-85eb-cfe79e6a1a8b",
   "metadata": {},
   "source": [
    "#### Check other scenarios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "abcf685b-e8d4-4b60-8269-8b84b3e24e2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If not check if other genes are within certain distance (2kb)\n",
    "nearby_gene_cols = ['nearby_gene_within_2kb', 'nearby_gene_within_30kb']\n",
    "merged_df[nearby_gene_cols[0]] = merged_df.apply(find_nearby_genes, args=(genes_df_subset,2000), axis=1)\n",
    "merged_df[nearby_gene_cols[1]] = merged_df.apply(find_nearby_genes, args=(genes_df_subset,30000), axis=1)\n",
    "\n",
    "nearest_gene_cols = ['nearest_within2kb_gene_id',\n",
    "                     'nearest_within2kb_gene_name',\n",
    "                     'nearest_within2kb_gene_dist',\n",
    "                     'nearest_within2kb_nontarget_gene_id',\n",
    "                     'nearest_within2kb_nontarget_gene_name',\n",
    "                     'nearest_within2kb_nontarget_gene_dist']\n",
    "merged_df[nearest_gene_cols] = merged_df.apply(find_nearest_genes, args=(genes_df_subset,2000), axis=1)\n",
    "\n",
    "# Identify rows where the primary target was not found, but the closest nearby gene is within 2000bp.\n",
    "condition = (merged_df['target_gene_id'].isna()) & (merged_df['nearby_gene_within_2kb'].apply(len)!=0)\n",
    "\n",
    "# Set the flag and note for these re-assigned rows.\n",
    "merged_df.loc[condition, 'note'] = 'unique alignment, target gene mismatch'\n",
    "merged_df.loc[condition, 'flag'] = True\n",
    "\n",
    "# Check if near a cds\n",
    "cds_hit_condition = (\n",
    "    merged_df['target_gene_id'].isna() &\n",
    "    merged_df.apply(lambda row: is_near_cds_start(row['pos'], row['cds']), axis=1)\n",
    ")\n",
    "\n",
    "merged_df.loc[cds_hit_condition, 'note'] = 'unique alignment, aligned to a cds but far from tss'\n",
    "merged_df.loc[cds_hit_condition, 'flag'] = True\n",
    "\n",
    "# check whether within target gene body\n",
    "gene_body_condition = (\n",
    "    merged_df['target_gene_id'].isna() &\n",
    "    ~cds_hit_condition &\n",
    "    (merged_df['pos'] >= merged_df['gene_start']) &\n",
    "    (merged_df['pos'] <= merged_df['gene_end'])\n",
    ")\n",
    "merged_df.loc[gene_body_condition, 'note'] = 'unique alignment, within target gene body but far from any tss or cds'\n",
    "merged_df.loc[gene_body_condition, 'flag'] = True\n",
    "\n",
    "# Final flagging\n",
    "final_no_target_condition = merged_df['target_gene_id'].isna()\n",
    "merged_df.loc[final_no_target_condition, 'flag'] = True\n",
    "note_is_generic = merged_df['note'] == 'unique alignment'\n",
    "merged_df.loc[final_no_target_condition & note_is_generic, 'note'] = 'unique alignment, no target gene found'\n",
    "\n",
    "# Clean up temporary columns from original dataframes\n",
    "unique_sgrna_df_all.drop(columns=['chromosome_norm'], inplace=True)\n",
    "genes_df_subset.drop(columns=['chromosome_norm'], inplace=True)\n",
    "final_df = merged_df[list(unique_sgrna_df_all.columns) + primary_target_cols + nearby_gene_cols + nearest_gene_cols]\n",
    "unique_sgrna_df_all = final_df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "873dc421-b88f-428b-beca-1dbf1b987fcb",
   "metadata": {},
   "source": [
    "#### Annotate putative bidirectional promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "9cb381eb-075b-4c05-8141-633dc6f92828",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_sgrna_df_all['putative_bidirectional_promoter'] = False\n",
    "\n",
    "# If a sgRNA sits on a putative bidirectional promoter, there should be >=2 nearby genes \n",
    "more_than_2nearby = unique_sgrna_df_all['nearby_gene_within_2kb'].apply(lambda x: len(x)) > 1\n",
    "filtered = unique_sgrna_df_all[more_than_2nearby]\n",
    "\n",
    "# Prepare genes_df_subset\n",
    "genes_df_subset.set_index('gene_id', inplace=True)\n",
    "\n",
    "def gene_orientation(sgrna_pos, gene_list):\n",
    "    'Small helper function to determine nearby gene orientation'\n",
    "    left = []\n",
    "    right = []\n",
    "    for gene in gene_list:\n",
    "        if (genes_df_subset.loc[gene,'strand']=='-') & (genes_df_subset.loc[gene,'gene_start']<sgrna_pos):\n",
    "            left.append(gene)\n",
    "        elif (genes_df_subset.loc[gene,'strand']=='+') & (genes_df_subset.loc[gene,'gene_end']>sgrna_pos):\n",
    "            right.append(gene)\n",
    "    return left, right\n",
    "\n",
    "for index, row in filtered.iterrows():\n",
    "    left_gene, right_gene = gene_orientation(row['pos'], row['nearby_gene_within_2kb'])\n",
    "    # If there is a gene on the left pointing to left, and a gene on the right pointing to right\n",
    "    # sgRNA might be on a bidirectional promoter\n",
    "    if (len(left_gene)!=0) & (len(right_gene)!=0):\n",
    "        filtered.loc[index, 'putative_bidirectional_promoter'] = True\n",
    "\n",
    "unique_sgrna_df_all[more_than_2nearby] = filtered\n",
    "    \n",
    "# Revert genes_df_subset\n",
    "genes_df_subset.reset_index(names=['gene_id'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b808c6ac-44c0-48f0-bf42-28c4a7f22f32",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_sgrna_df_all.to_pickle('results/unique_sgrna_df_final.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "528c4f6d-644c-4f75-8626-3e1095fa9a02",
   "metadata": {},
   "source": [
    "### Annotate multiple aligned sgRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bb851baf-6712-4db6-aade-8ee76a1cea06",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add sgRNA information\n",
    "df1 = multi_sgrna_df_filtered[['sgRNA', 'chromosome', 'pos', 'seq', 'strand', 'PAM', 'note', 'flag']].copy()\n",
    "df1.rename(columns={'seq': 'seq_last19bp'}, inplace=True)\n",
    "df2 = sgRNA[['name', 'sgRNA', 'target_gene_name_from_sgRNA', 'gene_id', 'corrected_target_gene_name']].copy()\n",
    "df2.rename(columns={\n",
    "    'sgRNA': 'seq',\n",
    "    'gene_id': 'designed_target_gene_id',\n",
    "    'corrected_target_gene_name': 'designed_target_gene_name'\n",
    "}, inplace=True)\n",
    "merged_df = pd.merge(\n",
    "    df1,\n",
    "    df2,\n",
    "    left_on='sgRNA',\n",
    "    right_on='name',\n",
    "    how='left'  # 'left' merge keeps all records from the first DataFrame\n",
    ")\n",
    "merged_df.drop(columns='name', inplace=True)\n",
    "multi_sgrna_df_filtered = merged_df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e96891e-1882-4ae3-9045-97d09c374070",
   "metadata": {},
   "source": [
    "#### Check whether near target gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4b371747-6f91-4ad1-9844-839a052d9404",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check whether target gene is within certain distance (2kb)\n",
    "multi_sgrna_df_filtered['chromosome_norm'] = multi_sgrna_df_filtered['chromosome'].str.replace('chr', '')\n",
    "genes_df_subset['chromosome_norm'] = genes_df_subset['chromosome'].str.replace('chr', '')\n",
    "merged_df = pd.merge(\n",
    "    multi_sgrna_df_filtered,\n",
    "    genes_df_subset[['gene_id', 'gene_name', 'chromosome_norm', 'tss', 'cds', 'gene_start', 'gene_end']],\n",
    "    left_on=['designed_target_gene_id', 'chromosome_norm'],\n",
    "    right_on=['gene_id', 'chromosome_norm'],\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "primary_target_cols = ['target_gene_id', 'target_gene_name', 'distance_to_closest_target_tss']\n",
    "merged_df[primary_target_cols] = merged_df.apply(find_closest_target_info, axis=1)\n",
    "\n",
    "match_condition = merged_df['designed_target_gene_id']==merged_df['target_gene_id']\n",
    "merged_df.loc[match_condition, 'note'] = 'multi alignment, aligned to target gene'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4612838-0423-4b5d-b82e-78cfb3d9eaba",
   "metadata": {},
   "source": [
    "#### Check other scenarios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d491a2d-f345-4a6c-af49-930676eeea54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If not check if other genes are within certain distance (2kb)\n",
    "nearby_gene_cols = ['nearby_gene_within_2kb', 'nearby_gene_within_30kb']\n",
    "merged_df[nearby_gene_cols[0]] = merged_df.apply(find_nearby_genes, args=(genes_df_subset,2000), axis=1)\n",
    "merged_df[nearby_gene_cols[1]] = merged_df.apply(find_nearby_genes, args=(genes_df_subset,30000), axis=1)\n",
    "\n",
    "nearest_gene_cols = ['nearest_within2kb_gene_id',\n",
    "                     'nearest_within2kb_gene_name',\n",
    "                     'nearest_within2kb_gene_dist',\n",
    "                     'nearest_within2kb_nontarget_gene_id',\n",
    "                     'nearest_within2kb_nontarget_gene_name',\n",
    "                     'nearest_within2kb_nontarget_gene_dist']\n",
    "merged_df[nearest_gene_cols] = merged_df.apply(find_nearest_genes, args=(genes_df_subset,2000), axis=1)\n",
    "\n",
    "# Identify rows where the primary target was not found, but the closest nearby gene is within 2000bp.\n",
    "condition = (merged_df['target_gene_id'].isna()) & (merged_df['nearby_gene_within_2kb'].apply(len)!=0)\n",
    "merged_df.loc[condition, 'note'] = 'multi alignment, target gene mismatch'\n",
    "\n",
    "# Check if near a cds\n",
    "cds_hit_condition = (\n",
    "    merged_df['target_gene_id'].isna() &\n",
    "    merged_df.apply(lambda row: is_near_cds_start(row['pos'], row['cds']), axis=1))\n",
    "merged_df.loc[cds_hit_condition, 'note'] = 'multi alignment, aligned to a cds but far from tss'\n",
    "\n",
    "# check whether within target gene body\n",
    "gene_body_condition = (\n",
    "    merged_df['target_gene_id'].isna() &\n",
    "    ~cds_hit_condition &\n",
    "    (merged_df['pos'] >= merged_df['gene_start']) &\n",
    "    (merged_df['pos'] <= merged_df['gene_end']))\n",
    "merged_df.loc[gene_body_condition, 'note'] = 'multi alignment, within target gene body but far from any tss or cds'\n",
    "merged_df.loc[gene_body_condition, 'flag'] = True\n",
    "\n",
    "# Final flagging\n",
    "final_no_target_condition = merged_df['target_gene_id'].isna()\n",
    "merged_df.loc[final_no_target_condition, 'flag'] = True\n",
    "note_is_generic = merged_df['note'] == 'multi alignment'\n",
    "merged_df.loc[final_no_target_condition & note_is_generic, 'note'] = 'multi alignment, no target gene found'\n",
    "\n",
    "# Clean up temporary columns from original dataframes\n",
    "multi_sgrna_df_filtered.drop(columns=['chromosome_norm'], inplace=True)\n",
    "genes_df_subset.drop(columns=['chromosome_norm'], inplace=True)\n",
    "final_df = merged_df[list(multi_sgrna_df_filtered.columns) + primary_target_cols + nearby_gene_cols + nearest_gene_cols]\n",
    "multi_sgrna_df_filtered = final_df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f905f8b-5028-4b28-baf2-3a229d843bdc",
   "metadata": {},
   "source": [
    "#### Annotate bidirectional promoter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcb6ee6e-3701-4f2a-84d9-c1944a758894",
   "metadata": {},
   "outputs": [],
   "source": [
    "multi_sgrna_df_filtered['putative_bidirectional_promoter'] = False\n",
    "\n",
    "# If a sgRNA sits on a putative bidirectional promoter, there should be >=2 nearby genes \n",
    "more_than_2nearby = multi_sgrna_df_filtered['nearby_gene_within_2kb'].apply(lambda x: len(x)) > 1\n",
    "filtered = multi_sgrna_df_filtered[more_than_2nearby]\n",
    "\n",
    "# Prepare genes_df_subset\n",
    "genes_df_subset.set_index('gene_id', inplace=True)\n",
    "\n",
    "for index, row in filtered.iterrows():\n",
    "    left_gene, right_gene = gene_orientation(row['pos'], row['nearby_gene_within_2kb'])\n",
    "    # If there is a gene on the left pointing to left, and a gene on the right pointing to right\n",
    "    # sgRNA might be on a bidirectional promoter\n",
    "    if (len(left_gene)!=0) & (len(right_gene)!=0):\n",
    "        filtered.loc[index, 'putative_bidirectional_promoter'] = True\n",
    "\n",
    "multi_sgrna_df_filtered[more_than_2nearby] = filtered\n",
    "    \n",
    "# Revert genes_df_subset\n",
    "genes_df_subset.reset_index(names=['gene_id'], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc2d094d-73bd-44d0-b3b0-d853b899d7a9",
   "metadata": {},
   "source": [
    "#### Resolve multi alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "561ab9a1-851e-4362-99aa-293da9430154",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Resolve chrX and chrY multi alignment\n",
    "key_cols = ['sgRNA', 'pos', 'strand']\n",
    "duplicated_sgrna = multi_sgrna_df_filtered[multi_sgrna_df_filtered.duplicated('sgRNA', keep=False)]\n",
    "grouped = duplicated_sgrna.groupby(key_cols)\n",
    "\n",
    "resolved_rows = []\n",
    "unresolved_indices = multi_sgrna_df_filtered.index.tolist()\n",
    "\n",
    "for _, group in grouped:\n",
    "    # Check if there are exactly 2 rows, one for chrX and one for chrY\n",
    "    is_chrX = 'chrX' in group['chromosome'].values\n",
    "    is_chrY = 'chrY' in group['chromosome'].values\n",
    "    # The group must have exactly 2 rows, one for chrX and one for chrY\n",
    "    if len(group) == 2 and is_chrX and is_chrY:\n",
    "        row_x = group[group['chromosome'] == 'chrX'].iloc[0]\n",
    "        row_y = group[group['chromosome'] == 'chrY'].iloc[0]\n",
    "        new_row = row_x.copy()\n",
    "        new_row['chromosome'] = 'chrX/Y'\n",
    "        if pd.isnull(row_x['target_gene_id']) & pd.isnull(row_y['target_gene_id']):\n",
    "            new_row['note'] = 'multi alignment, no target gene found on chrX/Y'\n",
    "        else:\n",
    "            dist_x = row_x['distance_to_closest_target_tss']\n",
    "            dist_y = row_y['distance_to_closest_target_tss']\n",
    "            new_row['distance_to_closest_target_tss'] = max(dist_x, dist_y)\n",
    "            new_row['target_gene_id'] = row_x['designed_target_gene_id']\n",
    "            new_row['note'] = 'multi alignment, target gene on chrX/Y'\n",
    "            new_row['flag'] = False\n",
    "\n",
    "        new_row['nearby_gene_within_2kb'] = list(set(row_x['nearby_gene_within_2kb']) | set(row_y['nearby_gene_within_2kb']))\n",
    "        new_row['nearby_gene_within_30kb'] = list(set(row_x['nearby_gene_within_30kb']) | set(row_y['nearby_gene_within_30kb']))\n",
    "\n",
    "        if pd.isna(row_x['nearest_within2kb_gene_dist']):\n",
    "            new_row['nearest_within2kb_gene_id'] = row_y['nearest_within2kb_gene_id']\n",
    "            new_row['nearest_within2kb_gene_name'] = row_y['nearest_within2kb_gene_name']\n",
    "            new_row['nearest_within2kb_gene_dist'] = row_y['nearest_within2kb_gene_dist']\n",
    "        elif not (pd.isna(row_x['nearest_within2kb_gene_dist']) & pd.isna(row_y['nearest_within2kb_gene_dist'])):\n",
    "            new_row['nearest_within2kb_gene_id'] = group.loc[group['nearest_within2kb_gene_dist'].idxmin(), 'nearest_within2kb_gene_id']\n",
    "            new_row['nearest_within2kb_gene_name'] = group.loc[group['nearest_within2kb_gene_dist'].idxmin(), 'nearest_within2kb_gene_name']\n",
    "            new_row['nearest_within2kb_gene_dist'] = group.loc[group['nearest_within2kb_gene_dist'].idxmin(), 'nearest_within2kb_gene_dist']\n",
    "\n",
    "        if pd.isna(row_x['nearest_within2kb_gene_dist']):\n",
    "            new_row['nearest_within2kb_nontarget_gene_id'] = row_y['nearest_within2kb_nontarget_gene_id']\n",
    "            new_row['nearest_within2kb_nontarget_gene_name'] = row_y['nearest_within2kb_nontarget_gene_name']\n",
    "            new_row['nearest_within2kb_nontarget_gene_dist'] = row_y['nearest_within2kb_nontarget_gene_dist']\n",
    "        elif not (pd.isna(row_x['nearest_within2kb_nontarget_gene_dist']) & pd.isna(row_y['nearest_within2kb_nontarget_gene_dist'])):\n",
    "            new_row['nearest_within2kb_nontarget_gene_id'] = group.loc[group['nearest_within2kb_nontarget_gene_dist'].idxmin(), 'nearest_within2kb_nontarget_gene_id']\n",
    "            new_row['nearest_within2kb_nontarget_gene_name'] = group.loc[group['nearest_within2kb_nontarget_gene_dist'].idxmin(), 'nearest_within2kb_nontarget_gene_name']\n",
    "            new_row['nearest_within2kb_nontarget_gene_dist'] = group.loc[group['nearest_within2kb_nontarget_gene_dist'].idxmin(), 'nearest_within2kb_nontarget_gene_dist']\n",
    "\n",
    "        new_row['putative_bidirectional_promoter'] = (row_x['putative_bidirectional_promoter']) | (row_y['putative_bidirectional_promoter'])\n",
    "        \n",
    "        resolved_rows.append(new_row.to_frame().T)\n",
    "        \n",
    "        # Remove the original chrX and chrY rows from the list of unresolved indices\n",
    "        unresolved_indices.remove(row_x.name)\n",
    "        unresolved_indices.remove(row_y.name)\n",
    "\n",
    "unresolved_df = multi_sgrna_df_filtered.loc[unresolved_indices].copy()\n",
    "resolved_df = pd.concat(resolved_rows, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "c8af19b0-ce19-4520-be92-e4476b6c04bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Resolve other multi alignments\n",
    "final_resolved_rows = []\n",
    "unresolved_df['other_alignment_chromosome'] = {}\n",
    "unresolved_df['other_alignment_pos'] = {}\n",
    "\n",
    "for _, group in unresolved_df.groupby('sgRNA', group_keys=False):\n",
    "    primary_row_mask = group['note'] == 'multi alignment, aligned to target gene'\n",
    "    # If at least one alignment aligned to target gene, collapse the group, add info from other rows to the primary row\n",
    "    if primary_row_mask.any():\n",
    "        primary_row = group[primary_row_mask].iloc[[0]].copy() \n",
    "        other_rows = group[~group.index.isin(primary_row.index)]\n",
    "    else:\n",
    "        if group['target_gene_id'].isnull().all():\n",
    "            if group['nearby_gene_within_2kb'].apply(lambda x: isinstance(x, list) and len(x) != 0).any():\n",
    "                # If at least one alignment has a target_gene_id, filter for rows that have a valid target gene ID, \n",
    "                # find the index of the row with the smallest distance to TSS\n",
    "                rows_with_target = group.dropna(subset=['nearest_within2kb_gene_dist']) \n",
    "                best_row_index = rows_with_target['nearest_within2kb_gene_dist'].idxmin()\n",
    "                primary_row = group.loc[[best_row_index]].copy()\n",
    "                other_rows = group.drop(best_row_index).copy()\n",
    "                primary_row['note'] = 'multi alignment, target gene mismatch'\n",
    "            # Check other scenarios\n",
    "            elif (group['note']=='multi alignment, aligned to a cds but far from tss').any():\n",
    "                best_row_index = group[group['note'] == 'multi alignment, aligned to a cds but far from tss'].index[0]\n",
    "                primary_row = group.loc[[best_row_index]].copy()\n",
    "                other_rows = group.drop(best_row_index).copy()\n",
    "                primary_row['note'] = 'multi alignment, aligned to a cds but far from tss'\n",
    "            elif (group['note']=='multi alignment, within target gene body but far from any tss or cds').any():\n",
    "                best_row_index = group[group['note'] == 'multi alignment, within target gene body but far from any tss or cds'].index[0]\n",
    "                primary_row = group.loc[[best_row_index]].copy()\n",
    "                other_rows = group.drop(best_row_index).copy()\n",
    "                primary_row['note'] = 'multi alignment, within target gene body but far from any tss or cds'\n",
    "            else:\n",
    "                # If no alignment has any target gene, take the first row as the representative\n",
    "                primary_row = group.iloc[[0]].copy()\n",
    "                other_rows = group.iloc[1:].copy()\n",
    "                primary_row['note'] = 'multi alignment, no target gene found'\n",
    "\n",
    "\n",
    "    # Add new columns with information about the other alignments\n",
    "    primary_row.at[primary_row.index[0], 'other_alignment_chromosome'] = list(other_rows['chromosome'])\n",
    "    primary_row.at[primary_row.index[0], 'other_alignment_pos'] = list(other_rows['pos'])\n",
    "\n",
    "    primary_row.at[primary_row.index[0], 'nearby_gene_within_2kb'] = list(set(group['nearby_gene_within_2kb'].explode().dropna().tolist()))\n",
    "    primary_row.at[primary_row.index[0], 'nearby_gene_within_30kb'] = list(set(group['nearby_gene_within_30kb'].explode().dropna().tolist()))\n",
    "\n",
    "    filtered = group.dropna(subset=['nearest_within2kb_gene_dist'])\n",
    "    if not filtered.empty:\n",
    "        primary_row['nearest_within2kb_gene_id'] = filtered.loc[filtered['nearest_within2kb_gene_dist'].idxmin(), 'nearest_within2kb_gene_id']\n",
    "        primary_row['nearest_within2kb_gene_name'] = filtered.loc[filtered['nearest_within2kb_gene_dist'].idxmin(), 'nearest_within2kb_gene_name']\n",
    "        primary_row['nearest_within2kb_gene_dist'] = filtered.loc[filtered['nearest_within2kb_gene_dist'].idxmin(), 'nearest_within2kb_gene_dist']\n",
    "\n",
    "        filtered = group.dropna(subset=['nearest_within2kb_nontarget_gene_dist'])\n",
    "    if not filtered.empty:\n",
    "        primary_row['nearest_within2kb_nontarget_gene_id'] = filtered.loc[filtered['nearest_within2kb_nontarget_gene_dist'].idxmin(), 'nearest_within2kb_nontarget_gene_id']\n",
    "        primary_row['nearest_within2kb_nontarget_gene_name'] = filtered.loc[filtered['nearest_within2kb_nontarget_gene_dist'].idxmin(), 'nearest_within2kb_nontarget_gene_name']\n",
    "        primary_row['nearest_within2kb_nontarget_gene_dist'] = filtered.loc[filtered['nearest_within2kb_nontarget_gene_dist'].idxmin(), 'nearest_within2kb_nontarget_gene_dist']\n",
    "\n",
    "    primary_row['putative_bidirectional_promoter'] = group['putative_bidirectional_promoter'].any()\n",
    "    \n",
    "    final_resolved_rows.append(primary_row.iloc[0])\n",
    "\n",
    "final_resolved_df = pd.DataFrame(final_resolved_rows)\n",
    "multi_sgrna_df_final = pd.concat([resolved_df, final_resolved_df])\n",
    "multi_sgrna_df_final = multi_sgrna_df_final.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "54905ec5-4773-48a5-ae83-ea10b74fb516",
   "metadata": {},
   "outputs": [],
   "source": [
    "multi_sgrna_df_final.to_pickle('results/multi_sgrna_df_final.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "968879e6-cac9-4dd6-b6de-df010c30597c",
   "metadata": {},
   "source": [
    "### Spot checking 7 unaligned sequences, all aligned to regions that are different in hg17 and hg18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "095173c8-f2f6-4e58-9f68-34b64a785cb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "unaligned_sgrna_df = pd.DataFrame({\n",
    "    'sgRNA': ['FDXR-1','MUC12-1','TYW1B-1','NBPF14-1','MUC12-2','SRGAP2-2','NBPF11-2'],\n",
    "    'seq':['AGGTTGCTGTTCCCAGCCA',\n",
    "           'CCTCCTGACGAGGGAAGAC',\n",
    "           'GCGTCGTGCAGATACTAGT',\n",
    "           'AATATTTATCAAACAGAGG',\n",
    "           'TCTTGGTCCCTCCTGACGA',\n",
    "           'CACAGGTCCCGAGCCAACG',\n",
    "           'GCCTGCAGTCCTAAACGCT'],\n",
    "    'chromosome':['chr17','chr7','chr7','chr1','chr7','chr1','chr1'],\n",
    "    'pos':[74872943,100969575,72744713,120451417,100969567,206463364,148152266],\n",
    "    'strand':['-','+','-','+','+','+','-'],\n",
    "    'note':['hg17 alignment, CtoT at pos10',\n",
    "            'hg17 alignment, TtoC at pos9, AtoG at pos13, TtoC at pos19',\n",
    "            'hg17 alignment, AtoG at pos6',\n",
    "            'hg17 alignment, AtoG at pos19, multi alignment',\n",
    "            'hg17 alignment, TtoC at pos17',\n",
    "            'hg17 alignment, AtoG at pos11',\n",
    "            'hg17 alignment, GtoT at pos19',\n",
    "           ],\n",
    "    'PAM':[True, True, True, True, False, True, True],\n",
    "    'flag':[True, True, True, True, True, True, True]\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "384f09be-232e-4fcf-83b3-fa38f65e19e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add sgRNA information\n",
    "df1 = unaligned_sgrna_df[['sgRNA', 'chromosome', 'pos', 'seq', 'strand', 'PAM', 'note', 'flag']].copy()\n",
    "df1.rename(columns={'seq': 'seq_last19bp'}, inplace=True)\n",
    "df2 = sgRNA[['name', 'sgRNA', 'target_gene_name_from_sgRNA', 'gene_id', 'corrected_target_gene_name']].copy()\n",
    "df2.rename(columns={\n",
    "    'sgRNA': 'seq',\n",
    "    'gene_id': 'designed_target_gene_id',\n",
    "    'corrected_target_gene_name': 'designed_target_gene_name'\n",
    "}, inplace=True)\n",
    "merged_df = pd.merge(\n",
    "    df1,\n",
    "    df2,\n",
    "    left_on='sgRNA',\n",
    "    right_on='name',\n",
    "    how='left'  # 'left' merge keeps all records from the first DataFrame\n",
    ")\n",
    "merged_df.drop(columns='name', inplace=True)\n",
    "unaligned_sgrna_df = merged_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "803c9538-a54e-4677-b5ac-c0b0cafe967b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check whether target gene is within certain distance (2kb)\n",
    "unaligned_sgrna_df['chromosome_norm'] = unaligned_sgrna_df['chromosome'].str.replace('chr', '')\n",
    "genes_df_subset['chromosome_norm'] = genes_df_subset['chromosome'].str.replace('chr', '')\n",
    "merged_df = pd.merge(\n",
    "    unaligned_sgrna_df,\n",
    "    genes_df_subset[['gene_id', 'gene_name', 'chromosome_norm', 'tss', 'cds', 'gene_start', 'gene_end']],\n",
    "    left_on=['designed_target_gene_id', 'chromosome_norm'],\n",
    "    right_on=['gene_id', 'chromosome_norm'],\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "primary_target_cols = ['target_gene_id', 'target_gene_name', 'distance_to_closest_target_tss']\n",
    "merged_df[primary_target_cols] = merged_df.apply(find_closest_target_info, axis=1)\n",
    "\n",
    "match_condition = merged_df['designed_target_gene_id']==merged_df['target_gene_id']\n",
    "merged_df.loc[match_condition, 'note'] = merged_df.loc[match_condition, 'note'] + ', aligned to target gene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "7fc12127-6e29-43bd-b24c-dd4c74722087",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If not check if other genes are within certain distance (2kb)\n",
    "nearby_gene_cols = ['nearby_gene_within_2kb', 'nearby_gene_within_30kb']\n",
    "merged_df[nearby_gene_cols[0]] = merged_df.apply(find_nearby_genes, args=(genes_df_subset,2000), axis=1)\n",
    "merged_df[nearby_gene_cols[1]] = merged_df.apply(find_nearby_genes, args=(genes_df_subset,30000), axis=1)\n",
    "\n",
    "nearest_gene_cols = ['nearest_within2kb_gene_id',\n",
    "                     'nearest_within2kb_gene_name',\n",
    "                     'nearest_within2kb_gene_dist',\n",
    "                     'nearest_within2kb_nontarget_gene_id',\n",
    "                     'nearest_within2kb_nontarget_gene_name',\n",
    "                     'nearest_within2kb_nontarget_gene_dist']\n",
    "merged_df[nearest_gene_cols] = merged_df.apply(find_nearest_genes, args=(genes_df_subset,2000), axis=1)\n",
    "\n",
    "# Identify rows where the primary target was not found, but the closest nearby gene is within 2000bp.\n",
    "condition = (merged_df['target_gene_id'].isna()) & (merged_df['nearby_gene_within_2kb'].apply(len)!=0)\n",
    "\n",
    "# Set the flag and note for these re-assigned rows.\n",
    "merged_df.loc[condition, 'note'] = 'unique alignment, target gene mismatch'\n",
    "merged_df.loc[condition, 'flag'] = True\n",
    "\n",
    "# Check if near a cds\n",
    "cds_hit_condition = (\n",
    "    merged_df['target_gene_id'].isna() &\n",
    "    merged_df.apply(lambda row: is_near_cds_start(row['pos'], row['cds']), axis=1)\n",
    ")\n",
    "\n",
    "merged_df.loc[cds_hit_condition, 'note'] = merged_df.loc[cds_hit_condition, 'note'] + ', aligned to a cds but far from tss'\n",
    "merged_df.loc[cds_hit_condition, 'flag'] = True\n",
    "\n",
    "# check whether within target gene body\n",
    "gene_body_condition = (\n",
    "    merged_df['target_gene_id'].isna() &\n",
    "    ~cds_hit_condition &\n",
    "    (merged_df['pos'] >= merged_df['gene_start']) &\n",
    "    (merged_df['pos'] <= merged_df['gene_end'])\n",
    ")\n",
    "merged_df.loc[gene_body_condition, 'note'] = merged_df.loc[gene_body_condition, 'note'] + ', within target gene body but far from any tss or cds'\n",
    "merged_df.loc[gene_body_condition, 'flag'] = True\n",
    "\n",
    "# Clean up temporary columns from original dataframes\n",
    "unaligned_sgrna_df.drop(columns=['chromosome_norm'], inplace=True)\n",
    "genes_df_subset.drop(columns=['chromosome_norm'], inplace=True)\n",
    "final_df = merged_df[list(unaligned_sgrna_df.columns) + primary_target_cols + nearby_gene_cols + nearest_gene_cols]\n",
    "unaligned_sgrna_df = final_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "ddb2f2d4-bcfc-4a5d-8eb3-18c5ec20b5a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "unaligned_sgrna_df['putative_bidirectional_promoter'] = False\n",
    "\n",
    "# If a sgRNA sits on a putative bidirectional promoter, there should be >=2 nearby genes \n",
    "more_than_2nearby = unaligned_sgrna_df['nearby_gene_within_2kb'].apply(lambda x: len(x)) > 1\n",
    "filtered = unaligned_sgrna_df[more_than_2nearby]\n",
    "\n",
    "# Prepare genes_df_subset\n",
    "genes_df_subset.set_index('gene_id', inplace=True)\n",
    "\n",
    "for index, row in filtered.iterrows():\n",
    "    left_gene, right_gene = gene_orientation(row['pos'], row['nearby_gene_within_2kb'])\n",
    "    # If there is a gene on the left pointing to left, and a gene on the right pointing to right\n",
    "    # sgRNA might be on a bidirectional promoter\n",
    "    if (len(left_gene)!=0) & (len(right_gene)!=0):\n",
    "        filtered.loc[index, 'putative_bidirectional_promoter'] = True\n",
    "\n",
    "unaligned_sgrna_df[more_than_2nearby] = filtered\n",
    "    \n",
    "# Revert genes_df_subset\n",
    "genes_df_subset.reset_index(names=['gene_id'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "147430e9-0f04-4264-96cf-58b6b61d7b11",
   "metadata": {},
   "outputs": [],
   "source": [
    "unaligned_sgrna_df['other_alignment_chromosome'] = {}\n",
    "unaligned_sgrna_df['other_alignment_pos'] = {}\n",
    "unaligned_sgrna_df.at[3, 'note'] = unaligned_sgrna_df.at[3, 'note'] + ', no target gene found'\n",
    "unaligned_sgrna_df.at[3, 'other_alignment_chromosome'] = ['chr1', 'chr1']\n",
    "unaligned_sgrna_df.at[3, 'other_alignment_pos'] = [120807845, 120814130]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "062d0f11-2fe3-4c43-9078-5a01e63eecbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "unaligned_sgrna_df.to_pickle('results/unaligned_sgrna_df_final.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "bb1dd0e2-15f4-4e1c-a927-868957f71090",
   "metadata": {},
   "outputs": [],
   "source": [
    "sgrna_df_final = pd.concat([unique_sgrna_df_all, multi_sgrna_df_final, unaligned_sgrna_df])\n",
    "sgrna_df_final = sgrna_df_final.reset_index(drop=True)\n",
    "sgrna_df_final.to_pickle('results/sgrna_df_final.pkl')\n",
    "sgrna_df_final.to_csv('results/sgrna_df_final.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18d2674a-6d08-4817-b2e1-b9ff74435f89",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
