## Main figures

| Figure | Description | Source |
|--------|-------------|--------|
| Figure 1A | Coverage and QC metrics. | `1_preprocess/qc_final.ipynb` |
| Figure 1D | Knock-down efficiency estimates. | `1_preprocess/estimate_guide_effect.ipynb` |
| Figure 1E | Comparison with Arrayed RNA-seq screens. | `3_DE_analysis/ArrayedRNAseq_comparison.ipynb` |
| Figure 1F | Comparison with FACS screens. | `3_DE_analysis/FACS_comparison_full.ipynb` |
| Figure 1G | Power analysis. | `3_DE_analysis/power_analysis/power_analysis.ipynb` |
| Figure 2A-B | Overview of cytokine regualators. | `5_cytokine_regulators/cytokine_regulators_overview.ipynb` |
| Figure 2C-F | Cytokine regualators validation | `5_cytokine_regulators/cytokine_followup_regulator.ipynb` |
| Figure 3 | Functional clustering. | [here] |
| Figure 3 | Functional clustering. | [here] |
| Figure 4C | Th2/Th1 polarization signatures. | `4_polarization_signatures/polarization_signature.ipynb` |
| Figure 4D-H | Regulators of Th2/Th1 polarization | `4_polarization_signatures/pert2state_polarization.ipynb` |
| Figure 5B | Age-associated signatures. | `7_1k1k_analysis/1k1k_DE_analysis.ipynb` |
| Figure 5C-F | Regulators of Age-associated signatures. | `7_1k1k_analysis/1k1k_pert2state_model.ipynb` | 
| Figure 6 | Regulator-LoF burden correlation | `8_lymphocyte_counts_LoF/lymph_reg_burden_correlation.ipynb` | 
| Figure 7 | Autoimmune disease-associated genes | `6_functional_interaction/autoimmune_analysis/opentargets_autoimmune_analysis.ipynb` |

## Supplementary figures

| Figure | Description | Source |
|--------|-------------|--------|
| Supplementary Figure 1  | Study design and QC metrics. | `1_preprocess/qc_final.ipynb` |
| Supplementary Figure 2  | Knock-down efficiency estimates. | `1_preprocess/estimate_guide_effect.ipynb` |
| Supplementary Figure 3  | Cell state variation in non-targeting control CD4+ T cells. | `2_embedding/ntc_embedding.ipynb` |
| Supplementary Figure 4  | Quality control of perturbation effects quantified by differential expression | `3_DE_analysis/DE_calibration.ipynb`|
|  | | `3_DE_analysis/DE_results_analysis_full.ipynb`| # rename
|  | | `3_DE_analysis/guide_robustness/DE_by_guide_results.ipynb`|
| Supplementary Figure 5  | Robustness of perturbation effects across donors. | `3_DE_analysis/donor_robustness/donor_robustness_analysis.ipynb` |
| Supplementary Figure 6  | Trans-effects of genetic perturbations. | `3_DE_analysis/DE_results_figure.ipynb` | # rename
|  | | `3_DE_analysis/DE_results_analysis_full.ipynb`| # rename
| Supplementary Figure 7  | Comparison with trans-effects in K562 genome-wide screen. | `3_DE_analysis/K562_DE/K562_vs_CD4T_analysis.ipynb` |
| Supplementary Figure 8  | Power analysis for perturb-seq dataset. | `3_DE_analysis/power_analysis/power_analysis.ipynb` |
| Supplementary Figure 9  | Power analysis results for each perturbed gene | `3_DE_analysis/power_analysis/power_analysis.ipynb` |
| Supplementary Figure 10 | Overview of trans-effects on 30 cytokines. | `5_cytokine_regulators/cytokine_regulators_overview.ipynb` |
| Supplementary Figure 11 | Arrayed validation of IL10/IL21 regulation. | `5_cytokine_regulators/cytokine_followup_regulator.ipynb` |
| Supplementary Figure 12 | Gene Ontology enrichment of downstream genes and an unknown regulator cluster. |  |
| Supplementary Figure 13 | Condition specificity regulator clusters. | `6_functional_interaction/condition_specificity.ipynb` |
| Supplementary Figure 14 | Tissue-specific expression patterns of gene clusters with or without gene set enrichment. | `6_functional_interaction/tissue_specificity.ipynb` |
| Supplementary Figure 15 |  Predicting regulators of Th2/Th1 polarization. | `4_polarization_signatures/pert2state_polarization.ipynb` | # rename
|  | | `4_polarization_signatures/polarization_signature.ipynb` | 
| Supplementary Figure 16 |  Effect of predicted polarization regulators on Th1 and Th2 genes. | `4_polarization_signatures/pert2state_polarization.ipynb` |
| Supplementary Figure 17 | Predicting regulators of Th2/Th1 polarization across CD4+T conditions.  |  `4_polarization_signatures/pert2state_polarization.ipynb` |
| Supplementary Figure 18 | Differential expression analysis of age-associated genes in CD4+ T cell subsets. | `7_1k1k_analysis/1k1k_DE_analysis.ipynb` |
| Supplementary Figure 19 | Predicting regulators of CD4+T cell aging signatures. | `7_1k1k_analysis/1k1k_pert2state_model.ipynb` | #rename
| Supplementary Figure 20 | Predicted regulators of CD4+T cell aging signature. | `7_1k1k_analysis/1k1k_pert2state_model.ipynb` | #rename
| Supplementary Figure 21 | Regulator-burden correlation on lymphocyte traits. | `8_lymphocyte_counts_LoF/lymph_reg_burden_correlation.ipynb` | 
| Supplementary Figure 22 | GWAS gene enrichment analysis across regulator clusters. | `6_functional_interaction/autoimmune_analysis/opentargets_autoimmune_analysis.ipynb` |
| Supplementary Figure 23 | Enrichment of autoimmune disease GWAS genes across regulatory clusters and their downstream targets. |  `6_functional_interaction/autoimmune_analysis/opentargets_autoimmune_analysis.ipynb` |
| Supplementary Figure 24 | Effects of perturbations of autoimmune associated genes in regulator clusters. |  `6_functional_interaction/autoimmune_analysis/opentargets_autoimmune_analysis.ipynb` |
