{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Curate sgRNA library metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1576374/1607250415.py:12: DtypeWarning: Columns (32,34,38,40,50) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  hgnc_complete = pd.read_csv(url, sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import genomic_features as gf\n",
    "\n",
    "## -- Read Replogle 2022 library gene names + IDs to settle multimapping -- ##\n",
    "\n",
    "replogle_library_genes = pd.read_csv('../../../../data/wesvae-data/K562_gwps_raw_singlecell_01.obs.csv')[['gene', 'gene_id']].drop_duplicates()\n",
    "\n",
    "## -- Get deprecated HGNC gene names -- ##\n",
    "# Download current HGNC dataset with previous symbols\n",
    "url = \"https://ftp.ebi.ac.uk/pub/databases/genenames/out_of_date_hgnc/tsv/hgnc_complete_set.txt\"\n",
    "hgnc_complete = pd.read_csv(url, sep='\\t')\n",
    "\n",
    "## -- Get Ensembl gene reference -- ##\n",
    "\n",
    "ensdb = gf.ensembl.annotation(species=\"Hsapiens\", version=\"110\") # matching reference in cellranger v9.0.0\n",
    "genes = ensdb.genes()\n",
    "# Filter biotypes\n",
    "keep_biotype = ['protein_coding', 'lncRNA', 'IG_V_pseudogene',\n",
    "       'TR_V_pseudogene', 'IG_V_gene', 'snRNA',\n",
    "       'miRNA', 'snoRNA', 'rRNA_pseudogene', 'rRNA',\n",
    "       'TR_V_gene', 'Mt_tRNA', 'Mt_rRNA', 'IG_C_gene', 'IG_J_gene',\n",
    "       'TR_J_gene', 'TR_C_gene', 'TR_J_pseudogene', 'IG_D_gene',\n",
    "       'ribozyme', 'IG_C_pseudogene', 'TR_D_gene', \n",
    "       'IG_J_pseudogene', 'scRNA', 'scaRNA',\n",
    "       'sRNA', 'IG_pseudogene']\n",
    "filt_genes = genes[genes.gene_biotype.isin(keep_biotype)]\n",
    "# Filter chromosomes\n",
    "filt_genes = filt_genes[filt_genes['seq_name'].isin([str(x) for x in np.arange(0,24)] + ['X', 'Y'])]\n",
    "# Use one source\n",
    "filt_genes['source'] = filt_genes['description'].str.split('[').str[-1].str.split(' ').str[0]\n",
    "filt_genes = filt_genes[filt_genes['source'] != 'Source:NCBI'].copy()\n",
    "\n",
    "## -- Annotate gene IDs for targeted genes -- ##\n",
    "\n",
    "# Read raw sgRnA library metadata\n",
    "sgrna_library_metadata = pd.read_csv('../../metadata/LRZ002RHSgRNA.csv', index_col=0)\n",
    "\n",
    "# Sanitize excel problems\n",
    "sgrna_library_metadata['name'] = np.where(sgrna_library_metadata['name'] == '1-Jun', 'JUN-1', sgrna_library_metadata['name'])\n",
    "sgrna_library_metadata['name'] = np.where(sgrna_library_metadata['name'] == '2-Jun', 'JUN-2', sgrna_library_metadata['name'])\n",
    "# Annotate perturbed gene\n",
    "sgrna_library_metadata['perturbed_gene_name'] = sgrna_library_metadata['name'].str.split('-').str[0:-1].str.join('-')\n",
    "\n",
    "sgrna_library_metadata = sgrna_library_metadata.rename({'name':'sgrna_id'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "325\n",
      "152\n"
     ]
    }
   ],
   "source": [
    "# Merge with Ensembl gene reference\n",
    "sgrna_library_metadata = pd.merge(sgrna_library_metadata, filt_genes.rename({'gene_name':'perturbed_gene_name'}, axis=1)[['perturbed_gene_name', 'gene_id', 'gene_biotype', 'description', 'source']], how='left')\n",
    "\n",
    "## -- Missing gene IDs -- ##\n",
    "missing_gene_id = sgrna_library_metadata[sgrna_library_metadata['gene_id'].isna() & (~sgrna_library_metadata.perturbed_gene_name.isin(['NTC', 'ProbeNTC']))].perturbed_gene_name.unique()\n",
    "print(len(missing_gene_id))\n",
    "# Settle missing with ID in Replogle2022\n",
    "missing_replogle2022 = replogle_library_genes[replogle_library_genes['gene'].isin(missing_gene_id)].set_index('gene').to_dict()['gene_id']\n",
    "sgrna_library_metadata.loc[sgrna_library_metadata['perturbed_gene_name'].isin(missing_replogle2022.keys()), 'gene_id'] = [missing_replogle2022[x] for x in sgrna_library_metadata.loc[sgrna_library_metadata['perturbed_gene_name'].isin(missing_replogle2022.keys()), 'perturbed_gene_name']] \n",
    "\n",
    "missing_gene_id = sgrna_library_metadata[sgrna_library_metadata['gene_id'].isna() & (~sgrna_library_metadata.perturbed_gene_name.isin(['NTC', 'ProbeNTC']))].perturbed_gene_name.unique()\n",
    "print(len(missing_gene_id))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 149 matching genes\n",
      "7\n"
     ]
    }
   ],
   "source": [
    "# Search gene names in renamed genes from HGNC\n",
    "keep_cols = ['perturbed_gene_name', 'hgnc_id', 'symbol',\n",
    "        'alias_symbol', 'alias_name',\n",
    "       'prev_symbol', 'prev_name', 'entrez_id', 'ensembl_gene_id', 'vega_id', 'ucsc_id',\n",
    "       ]\n",
    "\n",
    "def check_aliases(alias_str, target_genes):\n",
    "    if pd.isna(alias_str):\n",
    "        return False\n",
    "    aliases = str(alias_str).split('|')\n",
    "    return any(alias in target_genes for alias in aliases)\n",
    "\n",
    "def find_matching_query_gene(row, missing_gene_id):\n",
    "    # Check direct symbol match\n",
    "    if row['symbol'] in missing_gene_id:\n",
    "        return row['symbol']\n",
    "    \n",
    "    # Check alias matches\n",
    "    if pd.notna(row['alias_symbol']):\n",
    "        aliases = str(row['alias_symbol']).split('|')\n",
    "        matching = [a for a in aliases if a in missing_gene_id]\n",
    "        if matching:\n",
    "            return matching[0]\n",
    "            \n",
    "    # Check previous symbol matches\n",
    "    if pd.notna(row['prev_symbol']):\n",
    "        prev_symbols = str(row['prev_symbol']).split('|')\n",
    "        matching = [p for p in prev_symbols if p in missing_gene_id]\n",
    "        if matching:\n",
    "            return matching[0]\n",
    "    \n",
    "    return None\n",
    "\n",
    "# missing_gene_id = sgrna_library_metadata['perturbed_gene_name'].unique()\n",
    "\n",
    "# Create mask for direct symbol matches\n",
    "symbol_mask = hgnc_complete.symbol.isin(missing_gene_id)\n",
    "\n",
    "# Create mask for alias symbol matches \n",
    "alias_mask = hgnc_complete.alias_symbol.apply(lambda x: check_aliases(x, missing_gene_id))\n",
    "\n",
    "# Create mask for previous symbol matches\n",
    "prev_mask = hgnc_complete.prev_symbol.apply(lambda x: check_aliases(x, missing_gene_id))\n",
    "\n",
    "# Combine masks and filter\n",
    "matching_genes = hgnc_complete[symbol_mask | alias_mask | prev_mask].copy()\n",
    "\n",
    "# Add query gene column\n",
    "matching_genes['perturbed_gene_name'] = matching_genes.apply(lambda x: find_matching_query_gene(x, missing_gene_id), axis=1)\n",
    "\n",
    "# Select final columns\n",
    "matching_genes = matching_genes[keep_cols]\n",
    "\n",
    "print(f\"Found {len(matching_genes)} matching genes\")\n",
    "\n",
    "# Update gene_id in sgrna_library_metadata based on matching_genes using merge\n",
    "sgrna_library_metadata = sgrna_library_metadata.merge(\n",
    "    matching_genes[['perturbed_gene_name', 'ensembl_gene_id']],\n",
    "    on='perturbed_gene_name',\n",
    "    how='left'\n",
    ")\n",
    "\n",
    "# Update gene_id with ensembl_gene_id where gene_id is null\n",
    "sgrna_library_metadata.loc[sgrna_library_metadata['gene_id'].isna(), 'gene_id'] = \\\n",
    "    sgrna_library_metadata.loc[sgrna_library_metadata['gene_id'].isna(), 'ensembl_gene_id']\n",
    "\n",
    "# Drop the temporary ensembl_gene_id column\n",
    "sgrna_library_metadata = sgrna_library_metadata.drop('ensembl_gene_id', axis=1)\n",
    "\n",
    "missing_gene_id = sgrna_library_metadata[sgrna_library_metadata['gene_id'].isna() & (~sgrna_library_metadata.perturbed_gene_name.isin(['NTC', 'ProbeNTC']))].perturbed_gene_name.unique()\n",
    "print(len(missing_gene_id))\n",
    "# Just use gene name for highly deprecated genes\n",
    "sgrna_library_metadata['gene_id'] = np.where(sgrna_library_metadata['perturbed_gene_name'].isin(missing_gene_id), sgrna_library_metadata['perturbed_gene_name'], sgrna_library_metadata['gene_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['AKAP17A', 'ASMTL', 'CD99', 'CRLF2', 'CSF2RA', 'GTPBP6', 'IL3RA', 'IL9R', 'MUM1', 'P2RY8', 'PLCXD1', 'POLR2J3', 'PPP2R3B', 'SEPT2', 'SHOX', 'SLC25A6', 'VAMP7', 'ZBED1']\n",
      "Remaining multi-mapping genes: ['CRLF2', 'CSF2RA', 'IL3RA', 'IL9R', 'MUM1', 'P2RY8', 'SEPT2']\n"
     ]
    }
   ],
   "source": [
    "# Check double mappings\n",
    "mappings = sgrna_library_metadata[['perturbed_gene_name', 'gene_id']].drop_duplicates()\n",
    "n_mapped_ids = mappings.groupby('perturbed_gene_name').size()\n",
    "multimapping_genes = mappings.groupby('perturbed_gene_name').first()[n_mapped_ids > 1].index.tolist()\n",
    "print(multimapping_genes)\n",
    "\n",
    "# Settle multimappers with ID in Replogle2022\n",
    "multimappers_replogle2022 = replogle_library_genes[replogle_library_genes['gene'].isin(multimapping_genes)].set_index('gene').to_dict()['gene_id']\n",
    "sgrna_library_metadata = sgrna_library_metadata[\n",
    "    ~((sgrna_library_metadata['perturbed_gene_name'].isin(multimappers_replogle2022.keys())) &\n",
    "    (~sgrna_library_metadata['gene_id'].isin(multimappers_replogle2022.values())))\n",
    "]\n",
    "\n",
    "mappings = sgrna_library_metadata[['perturbed_gene_name', 'gene_id']].drop_duplicates()\n",
    "n_mapped_ids = mappings.groupby('perturbed_gene_name').size()\n",
    "multimapping_genes = mappings.groupby('perturbed_gene_name').first()[n_mapped_ids > 1].index.tolist()\n",
    "print(f'Remaining multi-mapping genes: {multimapping_genes}')\n",
    "\n",
    "# Keep first entry for remaining multimapping genes\n",
    "sgrna_library_metadata = sgrna_library_metadata.groupby('sgrna_id').first().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read raw sgRnA library metadata\n",
    "raw_sgrna_library_metadata = pd.read_csv('../../metadata/LRZ002RHSgRNA.csv', index_col=0)\n",
    "d = sgrna_library_metadata[['perturbed_gene_name', 'gene_id']].drop_duplicates().set_index('perturbed_gene_name')['gene_id'].to_dict()\n",
    "assert len([k for k,v in d.items() if v is None and k != 'NTC' and k != \"ProbeNTC\"]) == 0, 'Missing IDs'\n",
    "assert all([len(v) == 1 for k,v in d.items() if k != 'NTC' and k != \"ProbeNTC\"]) == 0, 'Duplicate ID'\n",
    "\n",
    "raw_sgrnas = raw_sgrna_library_metadata['name'].unique()\n",
    "post_sgrna = sgrna_library_metadata['sgrna_id'].unique()\n",
    "assert len(raw_sgrnas) == len(post_sgrna), 'Lost some sgRNA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save\n",
    "sgrna_library_metadata = sgrna_library_metadata.drop(['description', 'source', 'gene_biotype', 'read', 'pattern', 'feature_type'], axis=1)\n",
    "sgrna_library_metadata = sgrna_library_metadata.rename({\"gene_id\":'perturbed_gene_id'}, axis=1)\n",
    "sgrna_library_metadata.to_csv('../../metadata/sgRNA_library_curated.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rapids_singlecell",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
