{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "edf75ab0-5449-4cd9-81e0-a80793fbbd1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "import numpy as np\n",
    "from sgRNAalign_util import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb668a83-0a8c-4e64-9fe1-3a62f88c31e8",
   "metadata": {},
   "source": [
    "### Prepare sgRNA alignment file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8f34e764-ed85-4df6-9d79-5fc6d8b46fdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading from 'genome/GWCD4CRISPRi_nontargeting_sgRNA_list.csv'...\n",
      "Successfully converted 992 records.\n",
      "FASTA file saved as 'genome/GWCD4CRISPRi_nontargeting_sgRNA_list.fa'\n"
     ]
    }
   ],
   "source": [
    "convert_csv_to_fasta('genome/GWCD4CRISPRi_nontargeting_sgRNA_list.csv', 'genome/GWCD4CRISPRi_nontargeting_sgRNA_list.fa')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "675f3490-47d4-4794-ab6c-076d956f7e20",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "bowtie2 -x /Users/rzhu/Gladstone Dropbox/Ronghui Zhu/GRNPerturbSeq/2_files/hg38_genome/hg38 \\\n",
    "        -f genome/GWCD4CRISPRi_noncoding_sgRNA_list.fa \\\n",
    "        -S results/GWCD4CRISPRi_noncoding_sgRNA_alignment.sam \\\n",
    "        --end-to-end \\\n",
    "        -D 20 \\\n",
    "        -R 3 \\\n",
    "        -L 5 \\\n",
    "        -N 0 \\\n",
    "        -k 100 \\\n",
    "        -i S,1,0.50 \\\n",
    "        --score-min \"C,0,0\" \\\n",
    "        --no-1mm-upfront \\\n",
    "        --no-unal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36196257-60db-4f00-a5ab-b3ee9aa039d0",
   "metadata": {},
   "source": [
    "992 reads; of these:\n",
    "  992 (100.00%) were unpaired; of these:\n",
    "    978 (98.59%) aligned 0 times\n",
    "    12 (1.21%) aligned exactly 1 time\n",
    "    2 (0.20%) aligned >1 times\n",
    "1.41% overall alignment rate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6b6fa5f-5469-4305-a421-a388d518ab2c",
   "metadata": {},
   "source": [
    "### Extract alignment info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a570970e-928c-4e14-ace8-89e1c603e705",
   "metadata": {},
   "outputs": [],
   "source": [
    "alignment_df = sam_to_dataframe('results/GWCD4CRISPRi_nontargeting_sgRNA_alignment.sam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "862ee123-ac74-4916-99ce-531ff621fbc1",
   "metadata": {},
   "outputs": [],
   "source": "# Check whether there are PAMs\nalignment_df['PAM'] = True\n\nfrom pyfaidx import Fasta\n\n# Load the reference genome\ntry:\n    genome = Fasta('../../../2_files/hg38_genome/hg38.fa')\nexcept FileNotFoundError:\n    print(\"Error: hg38.fa not found. Please download the reference genome.\")\n    exit()\n\n# Flag uniquely aligned sgRNA that has no PAM\nfor i in range(len(alignment_df)):\n    # Define the genomic coordinates\n    chromosome = alignment_df.loc[i, 'chromosome']\n    pos = alignment_df.loc[i, 'pos']\n    if alignment_df.loc[i, 'strand']=='+':\n        start_position = pos+19\n        end_position = start_position+3\n        sequence = genome[chromosome][start_position:end_position]\n    else:\n        start_position = pos-3\n        end_position = pos\n        sequence = genome[chromosome][start_position:end_position]\n        sequence = sequence.reverse.complement\n    \n    if (sequence.seq[1:]!='GG') and (sequence.seq[1:]!='gg') and (sequence.seq[1:]!='Gg' and (sequence.seq[1:]!='gG')):\n        alignment_df.loc[i, 'PAM'] = False"
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7f09e14b-bcbe-4fe8-a821-4109312a4377",
   "metadata": {},
   "outputs": [],
   "source": [
    "ntc_sgRNA_df = alignment_df[alignment_df.PAM].copy()\n",
    "ntc_sgRNA_df = ntc_sgRNA_df.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa9a6ca6-3454-4b6e-a2eb-16cbe56257eb",
   "metadata": {},
   "source": [
    "## check whether there are genes nearby (2kb, 30kb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0537f57f-6261-4495-a0d6-f00c8f9c0c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_df_subset = pd.read_parquet('genome/genes_df_subset_for_sgRNA_annotation.parquet')\n",
    "ntc_sgRNA_df['chromosome_norm'] = ntc_sgRNA_df['chromosome'].str.replace('chr', '')\n",
    "genes_df_subset['chromosome_norm'] = genes_df_subset['chromosome'].str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "6f930109-bbd7-48b1-a84f-92ae9b76d3b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>pos</th>\n",
       "      <th>seq</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAM</th>\n",
       "      <th>chromosome_norm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NTC-424</td>\n",
       "      <td>chr10</td>\n",
       "      <td>109193861</td>\n",
       "      <td>CAGGATCAGGGTGTATGGC</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sgRNA chromosome        pos                  seq strand   PAM  \\\n",
       "0  NTC-424      chr10  109193861  CAGGATCAGGGTGTATGGC      +  True   \n",
       "\n",
       "  chromosome_norm  \n",
       "0              10  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ntc_sgRNA_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5316844c-1a6f-4cac-b231-2f2e403f56ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "dist1 = 2000\n",
    "dist2 = 30000\n",
    "for i, row in ntc_sgRNA_df.iterrows():\n",
    "    # Find nearest genes within certain distance on the same chromosome.\n",
    "    sgrna_chrom = row['chromosome_norm']\n",
    "    sgrna_pos = row['pos']\n",
    "    \n",
    "    same_chrom_genes = genes_df_subset[genes_df_subset['chromosome_norm'] == sgrna_chrom].copy()\n",
    "\n",
    "    # Calculate minimum distance for each gene\n",
    "    same_chrom_genes['distance'] = same_chrom_genes['tss'].apply(\n",
    "        lambda tss_list: min([abs(sgrna_pos - tss) for tss in tss_list])\n",
    "    )\n",
    "\n",
    "    # Filter by distance\n",
    "    nearby_genes_1 = same_chrom_genes[same_chrom_genes['distance'] <= dist1]\n",
    "    nearby_genes_2 = same_chrom_genes[same_chrom_genes['distance'] <= dist2]\n",
    "\n",
    "    if nearby_genes_1.empty:\n",
    "        nearest_gene_1_id = np.nan\n",
    "        nearest_gene_1_name = np.nan\n",
    "        nearest_gene_1_dist = np.nan\n",
    "    else:\n",
    "        nearest_gene_1_id = nearby_1_genes.loc[nearby_1_genes['distance'].idxmin(), 'gene_id']\n",
    "        nearest_gene_1_name = nearby_1_genes.loc[nearby_1_genes['distance'].idxmin(), 'gene_name']\n",
    "        nearest_gene_1_dist = nearby_1_genes.loc[nearby_1_genes['distance'].idxmin(), 'distance']\n",
    "\n",
    "    if nearby_genes_2.empty:\n",
    "        nearest_gene_2_id = np.nan\n",
    "        nearest_gene_2_name = np.nan\n",
    "        nearest_gene_2_dist = np.nan\n",
    "    else:\n",
    "        nearest_gene_2_id = nearby_2_genes.loc[nearby_2_genes['distance'].idxmin(), 'gene_id']\n",
    "        nearest_gene_2_name = nearby_2_genes.loc[nearby_2_genes['distance'].idxmin(), 'gene_name']\n",
    "        nearest_gene_2_dist = nearby_2_genes.loc[nearby_2_genes['distance'].idxmin(), 'distance']\n",
    "\n",
    "    ntc_sgRNA_df.loc[i, 'nearest_within2kb_gene_id'] = nearest_gene_1_id\n",
    "    ntc_sgRNA_df.loc[i, 'nearest_within2kb_gene_name'] = nearest_gene_1_name\n",
    "    ntc_sgRNA_df.loc[i, 'nearest_within2kb_gene_dist'] = nearest_gene_1_dist\n",
    "    ntc_sgRNA_df.loc[i, 'nearest_within30kb_gene_id'] = nearest_gene_2_id\n",
    "    ntc_sgRNA_df.loc[i, 'nearest_within30kb_gene_name'] = nearest_gene_2_name\n",
    "    ntc_sgRNA_df.loc[i, 'nearest_within30kb_gene_dist'] = nearest_gene_2_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "aab3a2d0-5ea3-46f1-8456-57084dfb3b13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>pos</th>\n",
       "      <th>seq</th>\n",
       "      <th>strand</th>\n",
       "      <th>PAM</th>\n",
       "      <th>chromosome_norm</th>\n",
       "      <th>nearest_within2kb_gene_id</th>\n",
       "      <th>nearest_within2kb_gene_name</th>\n",
       "      <th>nearest_within2kb_gene_dist</th>\n",
       "      <th>nearest_within30kb_gene_id</th>\n",
       "      <th>nearest_within30kb_gene_name</th>\n",
       "      <th>nearest_within30kb_gene_dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NTC-424</td>\n",
       "      <td>chr10</td>\n",
       "      <td>109193861</td>\n",
       "      <td>CAGGATCAGGGTGTATGGC</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     sgRNA chromosome        pos                  seq strand   PAM  \\\n",
       "0  NTC-424      chr10  109193861  CAGGATCAGGGTGTATGGC      +  True   \n",
       "\n",
       "  chromosome_norm  nearest_within2kb_gene_id  nearest_within2kb_gene_name  \\\n",
       "0              10                        NaN                          NaN   \n",
       "\n",
       "   nearest_within2kb_gene_dist  nearest_within30kb_gene_id  \\\n",
       "0                          NaN                         NaN   \n",
       "\n",
       "   nearest_within30kb_gene_name  nearest_within30kb_gene_dist  \n",
       "0                           NaN                           NaN  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ntc_sgRNA_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}