{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b2df04f9-4030-4cc1-9d4d-c5e5213e2066",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/rzhu/miniconda3/envs/deseq/lib/python3.11/site-packages/scanpy/_utils/__init__.py:33: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.\n",
      "  from anndata import __version__ as anndata_version\n",
      "/Users/rzhu/miniconda3/envs/deseq/lib/python3.11/site-packages/scanpy/__init__.py:24: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.\n",
      "  if Version(anndata.__version__) >= Version(\"0.11.0rc2\"):\n",
      "/Users/rzhu/miniconda3/envs/deseq/lib/python3.11/site-packages/scanpy/readwrite.py:16: FutureWarning: `__version__` is deprecated, use `importlib.metadata.version('anndata')` instead.\n",
      "  if Version(anndata.__version__) >= Version(\"0.11.0rc2\"):\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from joblib import Parallel, delayed\n",
    "import time\n",
    "import sys\n",
    "from tqdm import tqdm\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import hdbscan\n",
    "from scipy.sparse import csr_matrix\n",
    "from scipy.stats import hypergeom\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as mcolors \n",
    "import time\n",
    "\n",
    "\n",
    "sys.path.append('../3_DE_analysis/')\n",
    "from DE_analysis_utils import *\n",
    "\n",
    "pd.set_option('display.max_rows', 100)\n",
    "sc.set_figure_params(figsize=(20, 4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8cd7504f-01ef-460e-a2d4-a19c6194abc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_final = sc.read_h5ad('../../../../3_expts/processed_data/analysis_largefiles/nde30ntotal75_varfiltered_downstream_clustering_additional_filtering.h5ad', backed='r')\n",
    "de_by_guide = pd.read_csv('../3_DE_analysis/results/DE_by_guide.correlation_results.csv', index_col=0)\n",
    "donor_robustness_summary = pd.read_csv('../3_DE_analysis/results/Stim8hr_debyguidegenes_donor_robustness_summary.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "36d12564-00f2-4d58-a81f-7f56324de203",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_of_cluster = len(adata_final.obs['hdbscan'].unique())\n",
    "cluster_name = []\n",
    "corr = []\n",
    "cluster_gene_size = []\n",
    "cluster_member = []\n",
    "for i, cl in enumerate(adata_final.obs['hdbscan'].unique()):\n",
    "    df = pd.DataFrame(data=adata_final[adata_final.obs['hdbscan']==cl].layers['zscore'],\n",
    "                        index=adata_final[adata_final.obs['hdbscan']==cl].obs_names,\n",
    "                        columns=adata_final.var_names)\n",
    "    df_corr = df.T.corr()\n",
    "    np.fill_diagonal(df_corr.values, 0)\n",
    "    cluster_name.append(int(cl))\n",
    "    corr.append(np.mean(df_corr))\n",
    "    cluster_gene_size.append(len(df))\n",
    "    cluster_member.append(adata_final[adata_final.obs['hdbscan']==cl].obs.gene_name.tolist())\n",
    "\n",
    "cluster_df = pd.DataFrame({'cluster': cluster_name,\n",
    "                           'intracluster_corr': corr,\n",
    "                           'cluster_gene_size': cluster_gene_size,\n",
    "                           'cluster_member': cluster_member})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "89b4bee8-96ed-445c-93e1-b9e854f06876",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy\n",
    "from gseapy import Msigdb\n",
    "from gseapy import barplot, dotplot\n",
    "\n",
    "msig = Msigdb()\n",
    "kegg_gene_sets = msig.get_gmt(category= 'c2.cp.kegg_legacy', dbver=\"2025.1.Hs\")\n",
    "reactome_gene_sets = msig.get_gmt(category= 'c2.cp.reactome', dbver=\"2025.1.Hs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ca878427-870a-460c-810e-5c44d248695c",
   "metadata": {},
   "outputs": [],
   "source": [
    "corum_df = pd.read_csv('../../../../2_files/enrichment_analysis/CORUM/corum_humanComplexes.txt', delimiter='\\t', index_col='complex_id')\n",
    "stringdb = pd.read_csv('../../../../2_files/enrichment_analysis/STRINGDB/9606.clusters.proteins.v12.0.txt.gz', delimiter='\\t', compression='gzip')\n",
    "protein_info = pd.read_csv('../../../../2_files/enrichment_analysis/STRINGDB/9606.protein.info.v12.0.txt.gz', delimiter='\\t', compression='gzip')\n",
    "cluster_info = pd.read_csv('../../../../2_files/enrichment_analysis/STRINGDB/9606.clusters.info.v12.0.txt.gz', delimiter='\\t', compression='gzip')\n",
    "stringdb_df = pd.merge(stringdb, protein_info, left_on='protein_id', right_on='#string_protein_id')\n",
    "stringdb_df = pd.merge(stringdb_df, cluster_info, left_on='cluster_id', right_on='cluster_id')\n",
    "stringdb_df = stringdb_df[stringdb_df.cluster_size<1000].copy()\n",
    "zhang_df = pd.read_csv('../../../../2_files/enrichment_analysis/Zhang_et_al_2025/final_predictions_80.tsv', delimiter='\\t', skiprows=28, header=0)\n",
    "\n",
    "corum_complexes = {}\n",
    "for _, row in corum_df.iterrows():\n",
    "    complex_name = row['complex_name']\n",
    "    subunits = set(row['subunits_gene_name'].split(';'))\n",
    "    corum_complexes[complex_name] = subunits\n",
    "\n",
    "stringdb_complexes = {}\n",
    "for cluster_id in stringdb_df.cluster_id.unique():\n",
    "    stringdb_complexes[cluster_id] = set(stringdb_df[stringdb_df.cluster_id==cluster_id].preferred_name)\n",
    "\n",
    "zhang_complexes = {}\n",
    "for gene in list(set(zhang_df[['Name1','Name2']].values.flatten())):\n",
    "    zhang_complexes[gene] = set(zhang_df[(zhang_df.Name1==gene)|(zhang_df.Name2==gene)][['Name1','Name2']].values.flatten())\n",
    "\n",
    "zhang_pairwise_complexes = {}\n",
    "for i, row in zhang_df.iterrows():\n",
    "    zhang_pairwise_complexes['complex '+str(i)] = set([row['Name1'], row['Name2']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9cb31fda-a248-4d76-81ad-2116a93f201c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def assess_complex_enrichment(df, complexes, cluster_label, gene_name_label):\n",
    "    \"\"\"\n",
    "    Assess pathway/complex enrichment\n",
    "    \"\"\"\n",
    "    de_genes_per_cluster = {}\n",
    "    clusters = df[cluster_label].unique()\n",
    "    for cluster in clusters:\n",
    "        cluster_genes = df[df[cluster_label]==cluster][gene_name_label]\n",
    "        de_genes_per_cluster[cluster] = set(cluster_genes)\n",
    "\n",
    "    N = len(df)\n",
    "    \n",
    "    best_result = []\n",
    "    all_result = []\n",
    "    for cluster, de_genes in de_genes_per_cluster.items():\n",
    "        n = len(de_genes)\n",
    "        best_result_for_cluster = None # Initialize a variable to track the best result for this cluster\n",
    "        for complex_name, subunits in complexes.items():\n",
    "            K = len(subunits)\n",
    "            # Find the overlap between DE genes and complex subunits\n",
    "            overlap = de_genes.intersection(subunits)\n",
    "            k = len(overlap)\n",
    "            \n",
    "            # Calculate the p-value\n",
    "            if (n > 0 and K > 0) and (k > 1): # Avoid division by zero or trivial cases\n",
    "                pval = hypergeom.sf(k - 1, N, K, n) # sf is survival function\n",
    "                \n",
    "                # Adjust for multiple testing (e.g., using Benjamini-Hochberg)\n",
    "                # This is a crucial step to avoid false positives\n",
    "                # Store raw p-values and correct later\n",
    "                current_result = {\n",
    "                    'cluster': cluster,\n",
    "                    'complex': complex_name,\n",
    "                    'overlap_genes': list(overlap),\n",
    "                    'overlap_fraction': len(overlap)/n,\n",
    "                    'raw_p_value': pval,\n",
    "                    'complex_size': K,\n",
    "                    'overlap_size': len(overlap),\n",
    "                    'cluster_size': n\n",
    "                }\n",
    "                \n",
    "                if (best_result_for_cluster is None) or (pval < best_result_for_cluster['raw_p_value']):\n",
    "                    best_result_for_cluster = current_result\n",
    "                if len(overlap) > 1:\n",
    "                    all_result.append(current_result)\n",
    "        \n",
    "        if best_result_for_cluster is not None:\n",
    "            if best_result_for_cluster['raw_p_value']<1:\n",
    "                best_result.append(best_result_for_cluster)\n",
    "    \n",
    "    # Convert results to a DataFrame for easy viewing\n",
    "    best_result_df = pd.DataFrame(best_result)\n",
    "    #best_result_df = best_result_df.sort_values(by='cluster').reset_index(drop=True)\n",
    "    all_result_df = pd.DataFrame(all_result)\n",
    "    #all_result_df = all_result_df.sort_values(by='cluster').reset_index(drop=True)\n",
    "\n",
    "    return all_result_df, best_result_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9b51b377-6a83-42ac-afa5-101655b621e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_enrichment_analysis(df, cluster_label, gene_name_label):\n",
    "    # STRINGDB enrichment\n",
    "    stringdb_enrichment_all, stringdb_enrichment_best = assess_complex_enrichment(df, stringdb_complexes, cluster_label, gene_name_label)\n",
    "    stringdb_enrichment_all = pd.merge(stringdb_enrichment_all, stringdb_df[['cluster_id', 'best_described_by']].drop_duplicates(), left_on='complex', right_on='cluster_id')\n",
    "    stringdb_enrichment_all = stringdb_enrichment_all.drop(columns=['cluster_id'])\n",
    "    stringdb_enrichment_best = pd.merge(stringdb_enrichment_best, stringdb_df[['cluster_id', 'best_described_by']].drop_duplicates(), left_on='complex', right_on='cluster_id')\n",
    "    stringdb_enrichment_best = stringdb_enrichment_best.drop(columns=['cluster_id'])\n",
    "    # Corum enrichment\n",
    "    corum_enrichment_all, corum_enrichment_best = assess_complex_enrichment(df, corum_complexes, cluster_label, gene_name_label)\n",
    "    # KEGG enrichment\n",
    "    kegg_enrichment_all, kegg_enrichment_best = assess_complex_enrichment(df, kegg_gene_sets, cluster_label, gene_name_label)\n",
    "    # Reactome enrichment\n",
    "    reactome_enrichment_all, reactome_enrichment_best = assess_complex_enrichment(df, reactome_gene_sets, cluster_label, gene_name_label)\n",
    "    # Zhang et al 2025 enrichment\n",
    "    zhang_enrichment_all, zhang_enrichment_best = assess_complex_enrichment(df, zhang_complexes, cluster_label, gene_name_label)\n",
    "\n",
    "    # Zhang et al 2025 enrichment\n",
    "    zhang_pairwise_enrichment_all, zhang_pairwise_enrichment_best = assess_complex_enrichment(df, zhang_pairwise_complexes, cluster_label, gene_name_label)\n",
    "    \n",
    "    # Also checking inter-cluster overlap\n",
    "    gene_grouping = {}\n",
    "    gene_grouping_df = {}\n",
    "    for cluster in df[cluster_label].unique():\n",
    "        gene_grouping[cluster] = set(df[df[cluster_label]==cluster][gene_name_label])\n",
    "        gene_grouping_df[cluster] = [set(df[df[cluster_label]==cluster][gene_name_label])]\n",
    "    gene_grouping_df = pd.DataFrame(gene_grouping_df).T.rename(columns={0:'cluster_member'})\n",
    "    gene_grouping_df['cluster'] = gene_grouping_df.index\n",
    "    \n",
    "    intercluster_enrichment_all, _ = assess_complex_enrichment(df, gene_grouping, cluster_label, gene_name_label)\n",
    "    intercluster_enrichment_all = intercluster_enrichment_all[intercluster_enrichment_all.cluster!=intercluster_enrichment_all.complex].copy()\n",
    "    intercluster_enrichment_all = intercluster_enrichment_all[intercluster_enrichment_all.raw_p_value<1e-2].copy()\n",
    "    intercluster_enrichment_all = intercluster_enrichment_all[intercluster_enrichment_all.overlap_fraction>=0.2].copy()\n",
    "    intercluster_enrichment_summary = {}\n",
    "    \n",
    "    for cluster in intercluster_enrichment_all.cluster.unique():\n",
    "        intercluster_enrichment_summary[cluster] = set(intercluster_enrichment_all[intercluster_enrichment_all.cluster==cluster].complex)\n",
    "    \n",
    "    intercluster_enrichment_summary_df = pd.DataFrame(intercluster_enrichment_summary.items(), columns=['cluster', 'related_cluster'])\n",
    "    \n",
    "    # Summarize results\n",
    "    enrichment_df1 = pd.merge(corum_enrichment_best, stringdb_enrichment_best, on='cluster', how='outer', suffixes=('_corum', '_stringdb'))\n",
    "    enrichment_df2 = pd.merge(kegg_enrichment_best, reactome_enrichment_best, on='cluster', how='outer', suffixes=('_kegg', '_reactome'))\n",
    "    enrichment_df3 = pd.merge(zhang_enrichment_best, zhang_pairwise_enrichment_best, on='cluster', how='outer', suffixes=('_zhang', '_zhangpairwise'))\n",
    "    enrichment_df = pd.merge(enrichment_df1, enrichment_df2, on='cluster', how='outer')\n",
    "    enrichment_df = pd.merge(enrichment_df, enrichment_df3, on='cluster', how='outer')\n",
    "    enrichment_df = pd.merge(enrichment_df, cluster_df[['cluster', 'intracluster_corr', 'cluster_gene_size']], on='cluster', how='outer')\n",
    "    enrichment_df = pd.merge(enrichment_df, gene_grouping_df, on='cluster', how='outer')\n",
    "    enrichment_df = pd.merge(enrichment_df, intercluster_enrichment_summary_df, on='cluster', how='outer')\n",
    "    \n",
    "    return enrichment_df, corum_enrichment_all, stringdb_enrichment_all, kegg_enrichment_all, reactome_enrichment_all, zhang_enrichment_all, zhang_pairwise_enrichment_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8241d28d-cb21-4853-a952-f75b1309d1c1",
   "metadata": {},
   "source": [
    "### Check downstream cluster enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1fc28647-d57f-4199-835f-e4a369beb861",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_downstream_df,\\\n",
    "corum_enrichment_downstream_all,\\\n",
    "stringdb_enrichment_downstream_all,\\\n",
    "kegg_enrichment_downstream_all,\\\n",
    "reactome_enrichment_downstream_all,\\\n",
    "zhang_enrichment_downstream_all,\\\n",
    "zhang_pairwise_enrichment_downstream_all = run_enrichment_analysis(adata_final.obs[['hdbscan', 'gene_name']], 'hdbscan', 'gene_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "29e3c219-d024-4bde-b3e0-b3e401837d49",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_downstream_df.to_parquet('results/clustering_nde30ntotal75_downstream_cluster.parquet')\n",
    "enrichment_downstream_df.to_csv('results/clustering_nde30ntotal75_downstream_cluster.csv')\n",
    "corum_enrichment_downstream_all.to_parquet('results/corum_enrichment_nde30ntotal75_downstream_cluster.parquet')\n",
    "corum_enrichment_downstream_all.to_csv('results/corum_enrichment_nde30ntotal75_downstream_cluster.csv')\n",
    "stringdb_enrichment_downstream_all.to_parquet('results/stringdb_enrichment_nde30ntotal75_downstream_cluster.parquet')\n",
    "stringdb_enrichment_downstream_all.to_csv('results/stringdb_enrichment_nde30ntotal75_downstream_cluster.csv')\n",
    "kegg_enrichment_downstream_all.to_parquet('results/kegg_enrichment_nde30ntotal75_downstream_cluster.parquet')\n",
    "kegg_enrichment_downstream_all.to_csv('results/kegg_enrichment_nde30ntotal75_downstream_cluster.csv')\n",
    "reactome_enrichment_downstream_all.to_parquet('results/reactome_enrichment_nde30ntotal75_downstream_cluster.parquet')\n",
    "reactome_enrichment_downstream_all.to_csv('results/reactome_enrichment_nde30ntotal75_downstream_cluster.csv')\n",
    "zhang_enrichment_downstream_all.to_parquet('results/zhang_enrichment_nde30ntotal75_downstream_cluster.parquet')\n",
    "zhang_enrichment_downstream_all.to_csv('results/zhang_enrichment_nde30ntotal75_downstream_cluster.csv')\n",
    "zhang_pairwise_enrichment_downstream_all.to_parquet('results/zhang_pairwise_enrichment_nde30ntotal75_downstream_cluster.parquet')\n",
    "zhang_pairwise_enrichment_downstream_all.to_csv('results/zhang_pairwise_enrichment_nde30ntotal75_downstream_cluster.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9be2e237-76a4-484a-ab18-251b7c54768c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b05a283-5500-4466-b131-432207384566",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b7bc3018-01ab-4471-9afc-eea4ce0bc802",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = ['AASDH','ADAT3','ARL6IP1','ATRX','C1D','CALCOCO2','CAMSAP1','CD2','CD247','CD28',\n",
    "        'COLGALT1','COPS9','CREBBP','CYB5R4','DNAJC8','DOLPP1','EIF1','EIF4G2','ELOB','EWSR1',\n",
    "        'FOXM1','GNE','GPD1L','GPI','GSS','HELT','HTT','ITK','KAT2B','KIF14','KMT5B','KNOP1',\n",
    "        'L2HGDH','LRP5','MALT1','MEN1','METTL23','MTX2','NDFIP2','NUDCD1','NUP188','NUP37','PCNX3',\n",
    "        'PEX5','PHF23','PLCG1','PPHLN1','RNF8','SENP1','SHOC2','SIK3','SOCS4','STAT5A','SZRD1',\n",
    "        'TEAD4','TMEM263','TMX1','TNRC6A','TXNDC9','UBIAD1','XRRA1','ZC3H10','ZNF384','ZNF837']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8542b6cd-7b66-4939-8e13-ea03680f5ea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_filtered = ['AASDH','ADAT3','ARL6IP1','ATRX','C1D','CALCOCO2','CAMSAP1','COLGALT1','COPS9','CYB5R4','DNAJC8','DOLPP1','EIF1','EIF4G2','ELOB','EWSR1',\n",
    "        'FOXM1','GNE','GPD1L','GPI','GSS','HELT','HTT','KAT2B','KIF14','KMT5B','KNOP1','L2HGDH','LRP5','MEN1','METTL23','MTX2','NDFIP2','NUDCD1','NUP188','NUP37','PCNX3',\n",
    "        'PEX5','PHF23','PPHLN1','RNF8','SHOC2','SIK3','SZRD1','TEAD4','TMEM263','TMX1','TNRC6A','TXNDC9','UBIAD1','XRRA1','ZC3H10','ZNF384','ZNF837']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2b412bf0-b89c-47f5-a005-d0872654b6fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df = pd.DataFrame({'cluster':[109]*len(test), 'gene': test})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a45b36e1-afa4-4f88-97bc-8eb354c2bfa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "assess_complex_enrichment(test_df, reactome_gene_sets, 'cluster', 'gene')[0].tail(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "372950c9-c19e-4d6d-a779-91f925161444",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_stringdb_df, _ = assess_complex_enrichment(test_df, stringdb_complexes, 'cluster', 'gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "5359e70d-8539-487d-a4b0-1208bd9736bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>complex</th>\n",
       "      <th>overlap_genes</th>\n",
       "      <th>overlap_fraction</th>\n",
       "      <th>raw_p_value</th>\n",
       "      <th>complex_size</th>\n",
       "      <th>overlap_size</th>\n",
       "      <th>cluster_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17472</td>\n",
       "      <td>[SOCS4, ITK, PLCG1]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>1.0</td>\n",
       "      <td>56</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17470</td>\n",
       "      <td>[SOCS4, ITK, PLCG1]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>1.0</td>\n",
       "      <td>62</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17473</td>\n",
       "      <td>[SOCS4, PLCG1]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>33</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:6606</td>\n",
       "      <td>[FOXM1, KIF14]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>61</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:6608</td>\n",
       "      <td>[FOXM1, KIF14]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>51</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:2582</td>\n",
       "      <td>[NUP188, NUP37]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>61</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:2583</td>\n",
       "      <td>[NUP188, NUP37]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>42</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:2587</td>\n",
       "      <td>[NUP188, NUP37]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>27</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:2584</td>\n",
       "      <td>[NUP188, NUP37]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:2589</td>\n",
       "      <td>[NUP188, NUP37]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:34108</td>\n",
       "      <td>[TMX1, TXNDC9]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>465</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:34109</td>\n",
       "      <td>[TMX1, TXNDC9]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>434</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:29789</td>\n",
       "      <td>[PCNX3, KNOP1]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>195</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:25742</td>\n",
       "      <td>[PHF23, CAMSAP1]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>977</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:25743</td>\n",
       "      <td>[PHF23, CAMSAP1]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>771</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:25744</td>\n",
       "      <td>[PHF23, CAMSAP1]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>744</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:25745</td>\n",
       "      <td>[PHF23, CAMSAP1]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>716</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:25746</td>\n",
       "      <td>[PHF23, CAMSAP1]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>700</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:25747</td>\n",
       "      <td>[PHF23, CAMSAP1]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>680</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:25748</td>\n",
       "      <td>[PHF23, CAMSAP1]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>664</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:20248</td>\n",
       "      <td>[CREBBP, LRP5]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>733</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:20249</td>\n",
       "      <td>[CREBBP, LRP5]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>704</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:20250</td>\n",
       "      <td>[CREBBP, LRP5]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>697</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:20251</td>\n",
       "      <td>[CREBBP, LRP5]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>685</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:20252</td>\n",
       "      <td>[CREBBP, LRP5]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>662</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:20253</td>\n",
       "      <td>[CREBBP, LRP5]</td>\n",
       "      <td>0.031250</td>\n",
       "      <td>NaN</td>\n",
       "      <td>592</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17317</td>\n",
       "      <td>[SOCS4, SIK3, ITK, PLCG1, SHOC2]</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>702</td>\n",
       "      <td>5</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17318</td>\n",
       "      <td>[SOCS4, SHOC2, ITK, PLCG1]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>539</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17319</td>\n",
       "      <td>[SOCS4, SHOC2, ITK, PLCG1]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>515</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17320</td>\n",
       "      <td>[SOCS4, SHOC2, ITK, PLCG1]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>438</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17321</td>\n",
       "      <td>[SOCS4, SHOC2, ITK, PLCG1]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>245</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17323</td>\n",
       "      <td>[SOCS4, SHOC2, ITK, PLCG1]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>237</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17324</td>\n",
       "      <td>[SOCS4, SHOC2, ITK, PLCG1]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>230</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17325</td>\n",
       "      <td>[SOCS4, ITK, PLCG1]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>171</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17326</td>\n",
       "      <td>[SOCS4, ITK, PLCG1]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>164</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17327</td>\n",
       "      <td>[SOCS4, ITK, PLCG1]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>154</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:17328</td>\n",
       "      <td>[SOCS4, ITK, PLCG1]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>133</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15526</td>\n",
       "      <td>[STAT5A, MALT1, CD247, CD2, CD28]</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>896</td>\n",
       "      <td>5</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15527</td>\n",
       "      <td>[STAT5A, MALT1, CD247, CD2, CD28]</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>839</td>\n",
       "      <td>5</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15528</td>\n",
       "      <td>[STAT5A, MALT1, CD247, CD2, CD28]</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>808</td>\n",
       "      <td>5</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15529</td>\n",
       "      <td>[STAT5A, MALT1, CD247, CD2, CD28]</td>\n",
       "      <td>0.078125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>733</td>\n",
       "      <td>5</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15530</td>\n",
       "      <td>[CD247, STAT5A, CD28, CD2]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>526</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15531</td>\n",
       "      <td>[CD247, STAT5A, CD28, CD2]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>469</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15532</td>\n",
       "      <td>[CD247, STAT5A, CD28, CD2]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>455</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15533</td>\n",
       "      <td>[CD247, STAT5A, CD28, CD2]</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>346</td>\n",
       "      <td>4</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15534</td>\n",
       "      <td>[CD247, CD28, CD2]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>203</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15535</td>\n",
       "      <td>[CD247, CD28, CD2]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>185</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15536</td>\n",
       "      <td>[CD247, CD28, CD2]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>170</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15537</td>\n",
       "      <td>[CD247, CD28, CD2]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>109</td>\n",
       "      <td>CL:15538</td>\n",
       "      <td>[CD247, CD28, CD2]</td>\n",
       "      <td>0.046875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>93</td>\n",
       "      <td>3</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster   complex                      overlap_genes  overlap_fraction  \\\n",
       "26       109  CL:17472                [SOCS4, ITK, PLCG1]          0.046875   \n",
       "27       109  CL:17470                [SOCS4, ITK, PLCG1]          0.046875   \n",
       "29       109  CL:17473                     [SOCS4, PLCG1]          0.031250   \n",
       "67       109   CL:6606                     [FOXM1, KIF14]          0.031250   \n",
       "68       109   CL:6608                     [FOXM1, KIF14]          0.031250   \n",
       "99       109   CL:2582                    [NUP188, NUP37]          0.031250   \n",
       "100      109   CL:2583                    [NUP188, NUP37]          0.031250   \n",
       "101      109   CL:2587                    [NUP188, NUP37]          0.031250   \n",
       "102      109   CL:2584                    [NUP188, NUP37]          0.031250   \n",
       "103      109   CL:2589                    [NUP188, NUP37]          0.031250   \n",
       "0        109  CL:34108                     [TMX1, TXNDC9]          0.031250   \n",
       "1        109  CL:34109                     [TMX1, TXNDC9]          0.031250   \n",
       "2        109  CL:29789                     [PCNX3, KNOP1]          0.031250   \n",
       "3        109  CL:25742                   [PHF23, CAMSAP1]          0.031250   \n",
       "4        109  CL:25743                   [PHF23, CAMSAP1]          0.031250   \n",
       "5        109  CL:25744                   [PHF23, CAMSAP1]          0.031250   \n",
       "6        109  CL:25745                   [PHF23, CAMSAP1]          0.031250   \n",
       "7        109  CL:25746                   [PHF23, CAMSAP1]          0.031250   \n",
       "8        109  CL:25747                   [PHF23, CAMSAP1]          0.031250   \n",
       "9        109  CL:25748                   [PHF23, CAMSAP1]          0.031250   \n",
       "10       109  CL:20248                     [CREBBP, LRP5]          0.031250   \n",
       "11       109  CL:20249                     [CREBBP, LRP5]          0.031250   \n",
       "12       109  CL:20250                     [CREBBP, LRP5]          0.031250   \n",
       "13       109  CL:20251                     [CREBBP, LRP5]          0.031250   \n",
       "14       109  CL:20252                     [CREBBP, LRP5]          0.031250   \n",
       "15       109  CL:20253                     [CREBBP, LRP5]          0.031250   \n",
       "16       109  CL:17317   [SOCS4, SIK3, ITK, PLCG1, SHOC2]          0.078125   \n",
       "17       109  CL:17318         [SOCS4, SHOC2, ITK, PLCG1]          0.062500   \n",
       "18       109  CL:17319         [SOCS4, SHOC2, ITK, PLCG1]          0.062500   \n",
       "19       109  CL:17320         [SOCS4, SHOC2, ITK, PLCG1]          0.062500   \n",
       "20       109  CL:17321         [SOCS4, SHOC2, ITK, PLCG1]          0.062500   \n",
       "21       109  CL:17323         [SOCS4, SHOC2, ITK, PLCG1]          0.062500   \n",
       "22       109  CL:17324         [SOCS4, SHOC2, ITK, PLCG1]          0.062500   \n",
       "23       109  CL:17325                [SOCS4, ITK, PLCG1]          0.046875   \n",
       "24       109  CL:17326                [SOCS4, ITK, PLCG1]          0.046875   \n",
       "25       109  CL:17327                [SOCS4, ITK, PLCG1]          0.046875   \n",
       "28       109  CL:17328                [SOCS4, ITK, PLCG1]          0.046875   \n",
       "30       109  CL:15526  [STAT5A, MALT1, CD247, CD2, CD28]          0.078125   \n",
       "31       109  CL:15527  [STAT5A, MALT1, CD247, CD2, CD28]          0.078125   \n",
       "32       109  CL:15528  [STAT5A, MALT1, CD247, CD2, CD28]          0.078125   \n",
       "33       109  CL:15529  [STAT5A, MALT1, CD247, CD2, CD28]          0.078125   \n",
       "34       109  CL:15530         [CD247, STAT5A, CD28, CD2]          0.062500   \n",
       "35       109  CL:15531         [CD247, STAT5A, CD28, CD2]          0.062500   \n",
       "36       109  CL:15532         [CD247, STAT5A, CD28, CD2]          0.062500   \n",
       "37       109  CL:15533         [CD247, STAT5A, CD28, CD2]          0.062500   \n",
       "38       109  CL:15534                 [CD247, CD28, CD2]          0.046875   \n",
       "39       109  CL:15535                 [CD247, CD28, CD2]          0.046875   \n",
       "40       109  CL:15536                 [CD247, CD28, CD2]          0.046875   \n",
       "41       109  CL:15537                 [CD247, CD28, CD2]          0.046875   \n",
       "42       109  CL:15538                 [CD247, CD28, CD2]          0.046875   \n",
       "\n",
       "     raw_p_value  complex_size  overlap_size  cluster_size  \n",
       "26           1.0            56             3            64  \n",
       "27           1.0            62             3            64  \n",
       "29           1.0            33             2            64  \n",
       "67           1.0            61             2            64  \n",
       "68           1.0            51             2            64  \n",
       "99           1.0            61             2            64  \n",
       "100          1.0            42             2            64  \n",
       "101          1.0            27             2            64  \n",
       "102          1.0            34             2            64  \n",
       "103          1.0            18             2            64  \n",
       "0            NaN           465             2            64  \n",
       "1            NaN           434             2            64  \n",
       "2            NaN           195             2            64  \n",
       "3            NaN           977             2            64  \n",
       "4            NaN           771             2            64  \n",
       "5            NaN           744             2            64  \n",
       "6            NaN           716             2            64  \n",
       "7            NaN           700             2            64  \n",
       "8            NaN           680             2            64  \n",
       "9            NaN           664             2            64  \n",
       "10           NaN           733             2            64  \n",
       "11           NaN           704             2            64  \n",
       "12           NaN           697             2            64  \n",
       "13           NaN           685             2            64  \n",
       "14           NaN           662             2            64  \n",
       "15           NaN           592             2            64  \n",
       "16           NaN           702             5            64  \n",
       "17           NaN           539             4            64  \n",
       "18           NaN           515             4            64  \n",
       "19           NaN           438             4            64  \n",
       "20           NaN           245             4            64  \n",
       "21           NaN           237             4            64  \n",
       "22           NaN           230             4            64  \n",
       "23           NaN           171             3            64  \n",
       "24           NaN           164             3            64  \n",
       "25           NaN           154             3            64  \n",
       "28           NaN           133             3            64  \n",
       "30           NaN           896             5            64  \n",
       "31           NaN           839             5            64  \n",
       "32           NaN           808             5            64  \n",
       "33           NaN           733             5            64  \n",
       "34           NaN           526             4            64  \n",
       "35           NaN           469             4            64  \n",
       "36           NaN           455             4            64  \n",
       "37           NaN           346             4            64  \n",
       "38           NaN           203             3            64  \n",
       "39           NaN           185             3            64  \n",
       "40           NaN           170             3            64  \n",
       "41           NaN            98             3            64  \n",
       "42           NaN            93             3            64  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_stringdb_df.sort_values(by='raw_p_value').head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "83b6801a-aacd-4f25-8365-2ee7d0270147",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>complex</th>\n",
       "      <th>overlap_genes</th>\n",
       "      <th>overlap_fraction</th>\n",
       "      <th>raw_p_value</th>\n",
       "      <th>complex_size</th>\n",
       "      <th>overlap_size</th>\n",
       "      <th>cluster_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 1000</td>\n",
       "      <td>[CD2, SZRD1]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 1200</td>\n",
       "      <td>[DNAJC8, ATRX]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 1271</td>\n",
       "      <td>[CD247, CD2]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 4819</td>\n",
       "      <td>[CD28, CD2]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 5031</td>\n",
       "      <td>[SHOC2, ATRX]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 5061</td>\n",
       "      <td>[CD247, CD28]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 7365</td>\n",
       "      <td>[CD247, CD28]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 11778</td>\n",
       "      <td>[RNF8, ATRX]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 12242</td>\n",
       "      <td>[CD28, CD2]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 14897</td>\n",
       "      <td>[CD28, CD2]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cluster        complex   overlap_genes  overlap_fraction  raw_p_value  \\\n",
       "0      109   complex 1000    [CD2, SZRD1]           0.03125          1.0   \n",
       "1      109   complex 1200  [DNAJC8, ATRX]           0.03125          1.0   \n",
       "2      109   complex 1271    [CD247, CD2]           0.03125          1.0   \n",
       "3      109   complex 4819     [CD28, CD2]           0.03125          1.0   \n",
       "4      109   complex 5031   [SHOC2, ATRX]           0.03125          1.0   \n",
       "5      109   complex 5061   [CD247, CD28]           0.03125          1.0   \n",
       "6      109   complex 7365   [CD247, CD28]           0.03125          1.0   \n",
       "7      109  complex 11778    [RNF8, ATRX]           0.03125          1.0   \n",
       "8      109  complex 12242     [CD28, CD2]           0.03125          1.0   \n",
       "9      109  complex 14897     [CD28, CD2]           0.03125          1.0   \n",
       "\n",
       "   complex_size  overlap_size  cluster_size  \n",
       "0             9             2            64  \n",
       "1            17             2            64  \n",
       "2            14             2            64  \n",
       "3             7             2            64  \n",
       "4            37             2            64  \n",
       "5             7             2            64  \n",
       "6             9             2            64  \n",
       "7            12             2            64  \n",
       "8             5             2            64  \n",
       "9             2             2            64  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assess_complex_enrichment(test_df, zhang_complexes, 'cluster', 'gene')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "f57507bf-6b93-4e42-b872-685e0a0a9052",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>complex</th>\n",
       "      <th>overlap_genes</th>\n",
       "      <th>overlap_fraction</th>\n",
       "      <th>raw_p_value</th>\n",
       "      <th>complex_size</th>\n",
       "      <th>overlap_size</th>\n",
       "      <th>cluster_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 11465</td>\n",
       "      <td>[CD247, CD28]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>109</td>\n",
       "      <td>complex 20678</td>\n",
       "      <td>[CD28, CD2]</td>\n",
       "      <td>0.03125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cluster        complex  overlap_genes  overlap_fraction  raw_p_value  \\\n",
       "0      109  complex 11465  [CD247, CD28]           0.03125          1.0   \n",
       "1      109  complex 20678    [CD28, CD2]           0.03125          1.0   \n",
       "\n",
       "   complex_size  overlap_size  cluster_size  \n",
       "0             2             2            64  \n",
       "1             2             2            64  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assess_complex_enrichment(test_df, zhang_pairwise_complexes, 'cluster', 'gene')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "acdcc3e6-564a-4da2-b8d6-dad499ef52b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CD2', 'CD2BP2', 'EAPP', 'FEM1B', 'MAU2', 'NAA30', 'RHOT2', 'SF3B4', 'SZRD1'}"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhang_complexes['complex 1000']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "4374dd86-242e-4ca5-947a-9469f215bbdc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein1</th>\n",
       "      <th>Protein2</th>\n",
       "      <th>Name1</th>\n",
       "      <th>Name2</th>\n",
       "      <th>RFprob</th>\n",
       "      <th>AFprob</th>\n",
       "      <th>CFprob</th>\n",
       "      <th>AFMprob</th>\n",
       "      <th>Source</th>\n",
       "      <th>PDBtemp</th>\n",
       "      <th>...</th>\n",
       "      <th>Locality2</th>\n",
       "      <th>Disease1</th>\n",
       "      <th>Disease2</th>\n",
       "      <th>Process1</th>\n",
       "      <th>Process2</th>\n",
       "      <th>Function1</th>\n",
       "      <th>Function2</th>\n",
       "      <th>Exact_templates</th>\n",
       "      <th>Ortho_templates</th>\n",
       "      <th>Homo_templates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>O95400</td>\n",
       "      <td>Q15427</td>\n",
       "      <td>CD2BP2</td>\n",
       "      <td>SF3B4</td>\n",
       "      <td>0.50830</td>\n",
       "      <td>0.007206</td>\n",
       "      <td>0.008575</td>\n",
       "      <td>0.6426</td>\n",
       "      <td>P</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>Nucleus,Spliceosome</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>mRNA processing,mRNA splicing</td>\n",
       "      <td>mRNA processing,mRNA splicing</td>\n",
       "      <td>CD2 antigen cytoplasmic tail-binding protein 2</td>\n",
       "      <td>Splicing factor 3B subunit 4</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10981</th>\n",
       "      <td>O95400</td>\n",
       "      <td>Q56P03</td>\n",
       "      <td>CD2BP2</td>\n",
       "      <td>EAPP</td>\n",
       "      <td>0.10260</td>\n",
       "      <td>0.897000</td>\n",
       "      <td>0.958500</td>\n",
       "      <td>0.4573</td>\n",
       "      <td>P</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>Cytoplasm,Nucleus</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>mRNA processing,mRNA splicing</td>\n",
       "      <td>none</td>\n",
       "      <td>CD2 antigen cytoplasmic tail-binding protein 2</td>\n",
       "      <td>E2F-associated phosphoprotein</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14297</th>\n",
       "      <td>O95400</td>\n",
       "      <td>Q7Z422</td>\n",
       "      <td>CD2BP2</td>\n",
       "      <td>SZRD1</td>\n",
       "      <td>0.48460</td>\n",
       "      <td>0.885300</td>\n",
       "      <td>0.790500</td>\n",
       "      <td>0.2507</td>\n",
       "      <td>S</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>mRNA processing,mRNA splicing</td>\n",
       "      <td>none</td>\n",
       "      <td>CD2 antigen cytoplasmic tail-binding protein 2</td>\n",
       "      <td>SUZ domain-containing protein 1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14337</th>\n",
       "      <td>O95400</td>\n",
       "      <td>Q9UK73</td>\n",
       "      <td>CD2BP2</td>\n",
       "      <td>FEM1B</td>\n",
       "      <td>0.31960</td>\n",
       "      <td>0.805700</td>\n",
       "      <td>0.957500</td>\n",
       "      <td>0.9930</td>\n",
       "      <td>S</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>Cytoplasm,Nucleus</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>mRNA processing,mRNA splicing</td>\n",
       "      <td>Apoptosis,Ubl conjugation pathway</td>\n",
       "      <td>CD2 antigen cytoplasmic tail-binding protein 2</td>\n",
       "      <td>Protein fem-1 homolog B</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19657</th>\n",
       "      <td>O95400</td>\n",
       "      <td>Q9Y6X3</td>\n",
       "      <td>CD2BP2</td>\n",
       "      <td>MAU2</td>\n",
       "      <td>0.83740</td>\n",
       "      <td>0.717300</td>\n",
       "      <td>0.861300</td>\n",
       "      <td>0.6807</td>\n",
       "      <td>S</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>Chromosome,Nucleus</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>mRNA processing,mRNA splicing</td>\n",
       "      <td>Cell cycle,Cell division,Chromosome partition,...</td>\n",
       "      <td>CD2 antigen cytoplasmic tail-binding protein 2</td>\n",
       "      <td>MAU2 chromatid cohesion factor homolog</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28058</th>\n",
       "      <td>O95400</td>\n",
       "      <td>Q8IXI1</td>\n",
       "      <td>CD2BP2</td>\n",
       "      <td>RHOT2</td>\n",
       "      <td>0.38330</td>\n",
       "      <td>0.847700</td>\n",
       "      <td>0.969700</td>\n",
       "      <td>0.7470</td>\n",
       "      <td>S</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>Membrane,Mitochondrion,Mitochondrion outer mem...</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>mRNA processing,mRNA splicing</td>\n",
       "      <td>none</td>\n",
       "      <td>CD2 antigen cytoplasmic tail-binding protein 2</td>\n",
       "      <td>Mitochondrial Rho GTPase 2</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28545</th>\n",
       "      <td>O95400</td>\n",
       "      <td>P06729</td>\n",
       "      <td>CD2BP2</td>\n",
       "      <td>CD2</td>\n",
       "      <td>0.11505</td>\n",
       "      <td>0.565000</td>\n",
       "      <td>0.744000</td>\n",
       "      <td>0.6523</td>\n",
       "      <td>P</td>\n",
       "      <td>exact</td>\n",
       "      <td>...</td>\n",
       "      <td>Cell membrane,Membrane</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>mRNA processing,mRNA splicing</td>\n",
       "      <td>Cell adhesion</td>\n",
       "      <td>CD2 antigen cytoplasmic tail-binding protein 2</td>\n",
       "      <td>T-cell surface antigen CD2</td>\n",
       "      <td>1l2z</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28832</th>\n",
       "      <td>O95400</td>\n",
       "      <td>Q147X3</td>\n",
       "      <td>CD2BP2</td>\n",
       "      <td>NAA30</td>\n",
       "      <td>0.59330</td>\n",
       "      <td>0.046000</td>\n",
       "      <td>0.090900</td>\n",
       "      <td>0.9536</td>\n",
       "      <td>P</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>Cytoplasm,Nucleus</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>mRNA processing,mRNA splicing</td>\n",
       "      <td>none</td>\n",
       "      <td>CD2 antigen cytoplasmic tail-binding protein 2</td>\n",
       "      <td>N-alpha-acetyltransferase 30</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows  28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Protein1 Protein2   Name1  Name2   RFprob    AFprob    CFprob  AFMprob  \\\n",
       "538     O95400   Q15427  CD2BP2  SF3B4  0.50830  0.007206  0.008575   0.6426   \n",
       "10981   O95400   Q56P03  CD2BP2   EAPP  0.10260  0.897000  0.958500   0.4573   \n",
       "14297   O95400   Q7Z422  CD2BP2  SZRD1  0.48460  0.885300  0.790500   0.2507   \n",
       "14337   O95400   Q9UK73  CD2BP2  FEM1B  0.31960  0.805700  0.957500   0.9930   \n",
       "19657   O95400   Q9Y6X3  CD2BP2   MAU2  0.83740  0.717300  0.861300   0.6807   \n",
       "28058   O95400   Q8IXI1  CD2BP2  RHOT2  0.38330  0.847700  0.969700   0.7470   \n",
       "28545   O95400   P06729  CD2BP2    CD2  0.11505  0.565000  0.744000   0.6523   \n",
       "28832   O95400   Q147X3  CD2BP2  NAA30  0.59330  0.046000  0.090900   0.9536   \n",
       "\n",
       "      Source PDBtemp  ...                                          Locality2  \\\n",
       "538        P    none  ...                                Nucleus,Spliceosome   \n",
       "10981      P    none  ...                                  Cytoplasm,Nucleus   \n",
       "14297      S    none  ...                                               none   \n",
       "14337      S    none  ...                                  Cytoplasm,Nucleus   \n",
       "19657      S    none  ...                                 Chromosome,Nucleus   \n",
       "28058      S    none  ...  Membrane,Mitochondrion,Mitochondrion outer mem...   \n",
       "28545      P   exact  ...                             Cell membrane,Membrane   \n",
       "28832      P    none  ...                                  Cytoplasm,Nucleus   \n",
       "\n",
       "      Disease1  Disease2                       Process1  \\\n",
       "538       none      none  mRNA processing,mRNA splicing   \n",
       "10981     none      none  mRNA processing,mRNA splicing   \n",
       "14297     none      none  mRNA processing,mRNA splicing   \n",
       "14337     none      none  mRNA processing,mRNA splicing   \n",
       "19657     none      none  mRNA processing,mRNA splicing   \n",
       "28058     none      none  mRNA processing,mRNA splicing   \n",
       "28545     none      none  mRNA processing,mRNA splicing   \n",
       "28832     none      none  mRNA processing,mRNA splicing   \n",
       "\n",
       "                                                Process2  \\\n",
       "538                        mRNA processing,mRNA splicing   \n",
       "10981                                               none   \n",
       "14297                                               none   \n",
       "14337                  Apoptosis,Ubl conjugation pathway   \n",
       "19657  Cell cycle,Cell division,Chromosome partition,...   \n",
       "28058                                               none   \n",
       "28545                                      Cell adhesion   \n",
       "28832                                               none   \n",
       "\n",
       "                                            Function1  \\\n",
       "538    CD2 antigen cytoplasmic tail-binding protein 2   \n",
       "10981  CD2 antigen cytoplasmic tail-binding protein 2   \n",
       "14297  CD2 antigen cytoplasmic tail-binding protein 2   \n",
       "14337  CD2 antigen cytoplasmic tail-binding protein 2   \n",
       "19657  CD2 antigen cytoplasmic tail-binding protein 2   \n",
       "28058  CD2 antigen cytoplasmic tail-binding protein 2   \n",
       "28545  CD2 antigen cytoplasmic tail-binding protein 2   \n",
       "28832  CD2 antigen cytoplasmic tail-binding protein 2   \n",
       "\n",
       "                                    Function2 Exact_templates Ortho_templates  \\\n",
       "538              Splicing factor 3B subunit 4            none            none   \n",
       "10981           E2F-associated phosphoprotein            none            none   \n",
       "14297         SUZ domain-containing protein 1            none            none   \n",
       "14337                 Protein fem-1 homolog B            none            none   \n",
       "19657  MAU2 chromatid cohesion factor homolog            none            none   \n",
       "28058              Mitochondrial Rho GTPase 2            none            none   \n",
       "28545              T-cell surface antigen CD2            1l2z            none   \n",
       "28832            N-alpha-acetyltransferase 30            none            none   \n",
       "\n",
       "      Homo_templates  \n",
       "538             none  \n",
       "10981           none  \n",
       "14297           none  \n",
       "14337           none  \n",
       "19657           none  \n",
       "28058           none  \n",
       "28545           none  \n",
       "28832           none  \n",
       "\n",
       "[8 rows x 28 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhang_df[zhang_df.Name1=='CD2BP2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "087c4886-017f-4bd4-b4dd-017a70aeea35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein1</th>\n",
       "      <th>Protein2</th>\n",
       "      <th>Name1</th>\n",
       "      <th>Name2</th>\n",
       "      <th>RFprob</th>\n",
       "      <th>AFprob</th>\n",
       "      <th>CFprob</th>\n",
       "      <th>AFMprob</th>\n",
       "      <th>Source</th>\n",
       "      <th>PDBtemp</th>\n",
       "      <th>...</th>\n",
       "      <th>Locality2</th>\n",
       "      <th>Disease1</th>\n",
       "      <th>Disease2</th>\n",
       "      <th>Process1</th>\n",
       "      <th>Process2</th>\n",
       "      <th>Function1</th>\n",
       "      <th>Function2</th>\n",
       "      <th>Exact_templates</th>\n",
       "      <th>Ortho_templates</th>\n",
       "      <th>Homo_templates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4940</th>\n",
       "      <td>P06733</td>\n",
       "      <td>Q7Z422</td>\n",
       "      <td>ENO1</td>\n",
       "      <td>SZRD1</td>\n",
       "      <td>0.6187</td>\n",
       "      <td>0.06683</td>\n",
       "      <td>0.0551</td>\n",
       "      <td>0.6416</td>\n",
       "      <td>P</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>Glycolysis,Plasminogen activation,Transcriptio...</td>\n",
       "      <td>none</td>\n",
       "      <td>Alpha-enolase</td>\n",
       "      <td>SUZ domain-containing protein 1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14297</th>\n",
       "      <td>O95400</td>\n",
       "      <td>Q7Z422</td>\n",
       "      <td>CD2BP2</td>\n",
       "      <td>SZRD1</td>\n",
       "      <td>0.4846</td>\n",
       "      <td>0.88530</td>\n",
       "      <td>0.7905</td>\n",
       "      <td>0.2507</td>\n",
       "      <td>S</td>\n",
       "      <td>none</td>\n",
       "      <td>...</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>mRNA processing,mRNA splicing</td>\n",
       "      <td>none</td>\n",
       "      <td>CD2 antigen cytoplasmic tail-binding protein 2</td>\n",
       "      <td>SUZ domain-containing protein 1</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "      <td>none</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows  28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Protein1 Protein2   Name1  Name2  RFprob   AFprob  CFprob  AFMprob  \\\n",
       "4940    P06733   Q7Z422    ENO1  SZRD1  0.6187  0.06683  0.0551   0.6416   \n",
       "14297   O95400   Q7Z422  CD2BP2  SZRD1  0.4846  0.88530  0.7905   0.2507   \n",
       "\n",
       "      Source PDBtemp  ... Locality2 Disease1  Disease2  \\\n",
       "4940       P    none  ...      none     none      none   \n",
       "14297      S    none  ...      none     none      none   \n",
       "\n",
       "                                                Process1 Process2  \\\n",
       "4940   Glycolysis,Plasminogen activation,Transcriptio...     none   \n",
       "14297                      mRNA processing,mRNA splicing     none   \n",
       "\n",
       "                                            Function1  \\\n",
       "4940                                    Alpha-enolase   \n",
       "14297  CD2 antigen cytoplasmic tail-binding protein 2   \n",
       "\n",
       "                             Function2 Exact_templates Ortho_templates  \\\n",
       "4940   SUZ domain-containing protein 1            none            none   \n",
       "14297  SUZ domain-containing protein 1            none            none   \n",
       "\n",
       "      Homo_templates  \n",
       "4940            none  \n",
       "14297           none  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhang_df[zhang_df.Name2=='SZRD1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "479d49da-7b46-4e30-a4b4-38c14f857b50",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
